--- pytorch+++ mindspore@@ -1,7 +1,10 @@+import mindspore as ms
+import mindspore.nn as msnn
+import mindspore.ops as msops
+import mindspore.mint as mint
+from mindspore.mint import nn, ops
 from enum import Enum
 from typing import Union
-
-import torch
 
 
 class Format(str, Enum):
@@ -38,7 +41,7 @@     return dim
 
 
-def nchw_to(x: torch.Tensor, fmt: Format):
+def nchw_to(x: ms.Tensor, fmt: Format):
     if fmt == Format.NHWC:
         x = x.permute(0, 2, 3, 1)
     elif fmt == Format.NLC:
@@ -48,7 +51,7 @@     return x
 
 
-def nhwc_to(x: torch.Tensor, fmt: Format):
+def nhwc_to(x: ms.Tensor, fmt: Format):
     if fmt == Format.NCHW:
         x = x.permute(0, 3, 1, 2)
     elif fmt == Format.NLC:
