--- pytorch+++ mindspore@@ -1,12 +1,16 @@+import mindspore as ms
+import mindspore.nn as msnn
+import mindspore.ops as msops
+import mindspore.mint as mint
+from mindspore.mint import nn, ops
 """ Median Pool
 Hacked together by / Copyright 2020 Ross Wightman
 """
-import torch.nn as nn
-import torch.nn.functional as F
+# import torch.nn as nn
 from .helpers import to_2tuple, to_4tuple
 
 
-class MedianPool2d(nn.Module):
+class MedianPool2d(msnn.Cell):
     """ Median pool (usable as median filter when stride=1) module.
 
     Args:
@@ -42,8 +46,8 @@             padding = self.padding
         return padding
 
-    def forward(self, x):
-        x = F.pad(x, self._padding(x), mode='reflect')
+    def construct(self, x):
+        x = nn.functional.pad(x, self._padding(x), mode = 'reflect')
         x = x.unfold(2, self.k[0], self.stride[0]).unfold(3, self.k[1], self.stride[1])
         x = x.contiguous().view(x.size()[:4] + (-1,)).median(dim=-1)[0]
         return x
