--- pytorch+++ mindspore@@ -1,4 +1,9 @@ #!/usr/bin/env python3
+import mindspore as ms
+import mindspore.nn as msnn
+import mindspore.ops as msops
+import mindspore.mint as mint
+from mindspore.mint import nn, ops
 """ Checkpoint Cleaning Script
 
 Takes training checkpoints with GPU tensors, optimizer state, extra dict keys, etc.
@@ -7,7 +12,7 @@ 
 Hacked together by / Copyright 2020 Ross Wightman (https://github.com/rwightman)
 """
-import torch
+# import torch
 import argparse
 import os
 import hashlib
@@ -15,7 +20,7 @@ import tempfile
 from timm.models import load_state_dict
 try:
-    import safetensors.torch
+    # import safetensors.torch
     _has_safetensors = True
 except ImportError:
     _has_safetensors = False
@@ -88,7 +93,7 @@             assert _has_safetensors, "`pip install safetensors` to use .safetensors"
             safetensors.torch.save_file(new_state_dict, temp_filename)
         else:
-            torch.save(new_state_dict, temp_filename)
+            torch.save(new_state_dict, temp_filename)  # 'torch.save' 未在映射表(api_mapping_out_excel.json)中找到，需手动确认;
 
         with open(temp_filename, 'rb') as f:
             sha_hash = hashlib.sha256(f.read()).hexdigest()
