--- pytorch+++ mindspore@@ -1,13 +1,13 @@ from functools import wraps
-import torch
 from torch import nn
 
 from vit_pytorch.vit import Attention
+from mindspore.mint import nn, ops
 
 def find_modules(nn_module, type):
     return [module for module in nn_module.modules() if isinstance(module, type)]
 
-class Recorder(nn.Module):
+class Recorder(nn.Cell):
     def __init__(self, vit, device = None):
         super().__init__()
         self.vit = vit
@@ -55,5 +55,5 @@         target_device = self.device if self.device is not None else img.device
         recordings = tuple(map(lambda t: t.to(target_device), self.recordings))
 
-        attns = torch.stack(recordings, dim = 1) if len(recordings) > 0 else None
+        attns = ops.stack(tensors = recordings, dim = 1) if len(recordings) > 0 else None  # 'torch.stack':没有对应的mindspore参数 'out';
         return pred, attns
