--- pytorch+++ mindspore@@ -1,3 +1,4 @@+from mindspore.mint import nn, ops
 from random import randrange
 
 import torch
@@ -20,14 +21,14 @@     return unpack(t, ps, pattern)[0]
 
 def l2norm(t):
-    return F.normalize(t, dim = -1, p = 2)
+    return nn.functional.normalize(input = t, p = 2, dim = -1)  # 'torch.nn.functional.normalize':没有对应的mindspore参数 'out';
 
 def dropout_layers(layers, dropout):
     if dropout == 0:
         return layers
 
     num_layers = len(layers)
-    to_drop = torch.zeros(num_layers).uniform_(0., 1.) < dropout
+    to_drop = ops.zeros(size = num_layers).uniform_(0., 1.) < dropout  # 'torch.zeros':没有对应的mindspore参数 'out';; 'torch.zeros':没有对应的mindspore参数 'layout';; 'torch.zeros':没有对应的mindspore参数 'device';; 'torch.zeros':没有对应的mindspore参数 'requires_grad';
 
     # make sure at least one layer makes it
     if all(to_drop):
@@ -50,7 +51,7 @@             init_eps = 1e-6
 
         self.fn = fn
-        self.scale = nn.Parameter(torch.full((dim,), init_eps))
+        self.scale = nn.Parameter(ops.full(size = (dim,), fill_value = init_eps))  # 'torch.full':没有对应的mindspore参数 'out';; 'torch.full':没有对应的mindspore参数 'layout';; 'torch.full':没有对应的mindspore参数 'device';; 'torch.full':没有对应的mindspore参数 'requires_grad';
 
     def forward(self, x, **kwargs):
         return self.fn(x, **kwargs) * self.scale
@@ -59,13 +60,13 @@     def __init__(self, dim, hidden_dim, dropout = 0.):
         super().__init__()
         self.net = nn.Sequential(
-            nn.LayerNorm(dim),
-            nn.Linear(dim, hidden_dim),
+            nn.LayerNorm(normalized_shape = dim),
+            nn.Linear(in_features = dim, out_features = hidden_dim),
             nn.GELU(),
-            nn.Dropout(dropout),
-            nn.Linear(hidden_dim, dim),
-            nn.Dropout(dropout)
-        )
+            nn.Dropout(p = dropout),
+            nn.Linear(in_features = hidden_dim, out_features = dim),
+            nn.Dropout(p = dropout)
+        )  # 'torch.nn.LayerNorm':没有对应的mindspore参数 'device';; 'torch.nn.Linear':没有对应的mindspore参数 'device';
     def forward(self, x):
         return self.net(x)
 
@@ -76,33 +77,33 @@         self.heads = heads
         self.scale = dim_head ** -0.5
 
-        self.norm = nn.LayerNorm(dim)
-        self.to_q = nn.Linear(dim, inner_dim, bias = False)
-        self.to_kv = nn.Linear(dim, inner_dim * 2, bias = False)
+        self.norm = nn.LayerNorm(normalized_shape = dim)  # 'torch.nn.LayerNorm':没有对应的mindspore参数 'device';
+        self.to_q = nn.Linear(in_features = dim, out_features = inner_dim, bias = False)  # 'torch.nn.Linear':没有对应的mindspore参数 'device';
+        self.to_kv = nn.Linear(in_features = dim, out_features = inner_dim * 2, bias = False)  # 'torch.nn.Linear':没有对应的mindspore参数 'device';
 
         self.attend = nn.Softmax(dim = -1)
-        self.dropout = nn.Dropout(dropout)
+        self.dropout = nn.Dropout(p = dropout)
 
         self.to_out = nn.Sequential(
-            nn.Linear(inner_dim, dim),
-            nn.Dropout(dropout)
-        )
+            nn.Linear(in_features = inner_dim, out_features = dim),
+            nn.Dropout(p = dropout)
+        )  # 'torch.nn.Linear':没有对应的mindspore参数 'device';
 
     def forward(self, x, context = None):
         h = self.heads
 
         x = self.norm(x)
-        context = x if not exists(context) else torch.cat((x, context), dim = 1)
+        context = x if not exists(context) else ops.cat(tensors = (x, context), dim = 1)  # 'torch.cat':没有对应的mindspore参数 'out';
 
         qkv = (self.to_q(x), *self.to_kv(context).chunk(2, dim = -1))
         q, k, v = map(lambda t: rearrange(t, 'b n (h d) -> b h n d', h = h), qkv)
 
-        sim = einsum('b h i d, b h j d -> b h i j', q, k) * self.scale
+        sim = ops.einsum(equation = 'b h i d, b h j d -> b h i j', operands = q) * self.scale
 
         attn = self.attend(sim)
         attn = self.dropout(attn)
 
-        out = einsum('b h i j, b h j d -> b h i d', attn, v)
+        out = ops.einsum(equation = 'b h i j, b h j d -> b h i d', operands = attn)
         out = rearrange(out, 'b h n d -> b n (h d)')
         return self.to_out(out)
 
@@ -111,19 +112,19 @@         super().__init__()
         inner_dim = dim_head * heads
         self.heads = heads
-        self.norm = nn.LayerNorm(dim)
-
-        self.to_qkv = nn.Linear(dim, inner_dim * 3, bias = False)
-
-        self.temperature = nn.Parameter(torch.ones(heads, 1, 1))
+        self.norm = nn.LayerNorm(normalized_shape = dim)  # 'torch.nn.LayerNorm':没有对应的mindspore参数 'device';
+
+        self.to_qkv = nn.Linear(in_features = dim, out_features = inner_dim * 3, bias = False)  # 'torch.nn.Linear':没有对应的mindspore参数 'device';
+
+        self.temperature = nn.Parameter(ops.ones(size = heads, dtype = 1))  # 'torch.ones':没有对应的mindspore参数 'out';; 'torch.ones':没有对应的mindspore参数 'layout';; 'torch.ones':没有对应的mindspore参数 'device';; 'torch.ones':没有对应的mindspore参数 'requires_grad';
 
         self.attend = nn.Softmax(dim = -1)
-        self.dropout = nn.Dropout(dropout)
+        self.dropout = nn.Dropout(p = dropout)
 
         self.to_out = nn.Sequential(
-            nn.Linear(inner_dim, dim),
-            nn.Dropout(dropout)
-        )
+            nn.Linear(in_features = inner_dim, out_features = dim),
+            nn.Dropout(p = dropout)
+        )  # 'torch.nn.Linear':没有对应的mindspore参数 'device';
 
     def forward(self, x):
         h = self.heads
@@ -136,12 +137,12 @@ 
         q, k = map(l2norm, (q, k))
 
-        sim = einsum('b h i n, b h j n -> b h i j', q, k) * self.temperature.exp()
+        sim = ops.einsum(equation = 'b h i n, b h j n -> b h i j', operands = q) * self.temperature.exp()
 
         attn = self.attend(sim)
         attn = self.dropout(attn)
 
-        out = einsum('b h i j, b h j n -> b h i n', attn, v)
+        out = ops.einsum(equation = 'b h i j, b h j n -> b h i n', operands = attn)
         out = rearrange(out, 'b h d n -> b n (h d)')
 
         out = unpack_one(out, ps, 'b * d')
@@ -154,14 +155,14 @@         padding = kernel_size // 2
 
         self.net = nn.Sequential(
-            nn.LayerNorm(dim),
+            nn.LayerNorm(normalized_shape = dim),
             Rearrange('b h w c -> b c h w'),
-            nn.Conv2d(dim, dim, kernel_size, padding = padding, groups = dim),
-            nn.BatchNorm2d(dim),
+            nn.Conv2d(in_channels = dim, out_channels = dim, kernel_size = kernel_size, padding = padding, groups = dim),
+            nn.BatchNorm2d(num_features = dim),
             nn.GELU(),
-            nn.Conv2d(dim, dim, kernel_size, padding = padding, groups = dim),
+            nn.Conv2d(in_channels = dim, out_channels = dim, kernel_size = kernel_size, padding = padding, groups = dim),
             Rearrange('b c h w -> b h w c'),
-        )
+        )  # 'torch.nn.LayerNorm':没有对应的mindspore参数 'device';; 'torch.nn.Conv2d':没有对应的mindspore参数 'device';; 'torch.nn.BatchNorm2d':没有对应的mindspore参数 'device';
 
     def forward(self, x):
         return self.net(x)
@@ -238,26 +239,26 @@ 
         self.to_patch_embedding = nn.Sequential(
             Rearrange('b c (h p1) (w p2) -> b h w (p1 p2 c)', p1 = patch_size, p2 = patch_size),
-            nn.LayerNorm(patch_dim),
-            nn.Linear(patch_dim, dim),
-            nn.LayerNorm(dim)
-        )
-
-        self.pos_embedding = nn.Parameter(torch.randn(1, num_patches, dim))
-        self.cls_token = nn.Parameter(torch.randn(dim))
-
-        self.dropout = nn.Dropout(emb_dropout)
+            nn.LayerNorm(normalized_shape = patch_dim),
+            nn.Linear(in_features = patch_dim, out_features = dim),
+            nn.LayerNorm(normalized_shape = dim)
+        )  # 'torch.nn.LayerNorm':没有对应的mindspore参数 'device';; 'torch.nn.Linear':没有对应的mindspore参数 'device';
+
+        self.pos_embedding = nn.Parameter(ops.randn(size = 1, generator = num_patches))  # 'torch.randn':没有对应的mindspore参数 'out';; 'torch.randn':没有对应的mindspore参数 'layout';; 'torch.randn':没有对应的mindspore参数 'device';; 'torch.randn':没有对应的mindspore参数 'requires_grad';; 'torch.randn':没有对应的mindspore参数 'pin_memory';
+        self.cls_token = nn.Parameter(ops.randn(size = dim))  # 'torch.randn':没有对应的mindspore参数 'out';; 'torch.randn':没有对应的mindspore参数 'layout';; 'torch.randn':没有对应的mindspore参数 'device';; 'torch.randn':没有对应的mindspore参数 'requires_grad';; 'torch.randn':没有对应的mindspore参数 'pin_memory';
+
+        self.dropout = nn.Dropout(p = emb_dropout)
 
         self.xcit_transformer = XCATransformer(dim, depth, heads, dim_head, mlp_dim, local_patch_kernel_size, dropout, layer_dropout)
 
-        self.final_norm = nn.LayerNorm(dim)
+        self.final_norm = nn.LayerNorm(normalized_shape = dim)  # 'torch.nn.LayerNorm':没有对应的mindspore参数 'device';
 
         self.cls_transformer = Transformer(dim, cls_depth, heads, dim_head, mlp_dim, dropout, layer_dropout)
 
         self.mlp_head = nn.Sequential(
-            nn.LayerNorm(dim),
-            nn.Linear(dim, num_classes)
-        )
+            nn.LayerNorm(normalized_shape = dim),
+            nn.Linear(in_features = dim, out_features = num_classes)
+        )  # 'torch.nn.LayerNorm':没有对应的mindspore参数 'device';; 'torch.nn.Linear':没有对应的mindspore参数 'device';
 
     def forward(self, img):
         x = self.to_patch_embedding(img)
