--- pytorch+++ mindspore@@ -1,9 +1,8 @@-import torch
 from torch import nn, einsum
-import torch.nn.functional as F
 
 from einops import rearrange, repeat
 from einops.layers.torch import Rearrange
+from mindspore.mint import nn, ops
 
 # helper methods
 
@@ -22,7 +21,7 @@ 
 # classes
 
-class Residual(nn.Module):
+class Residual(nn.Cell):
     def __init__(self, fn):
         super().__init__()
         self.fn = fn
@@ -30,59 +29,59 @@     def forward(self, x, **kwargs):
         return self.fn(x, **kwargs) + x
 
-class LayerNorm(nn.Module):
+class LayerNorm(nn.Cell):
     def __init__(self, dim, eps = 1e-5):
         super().__init__()
         self.eps = eps
-        self.g = nn.Parameter(torch.ones(1, dim, 1, 1))
-        self.b = nn.Parameter(torch.zeros(1, dim, 1, 1))
-
-    def forward(self, x):
-        var = torch.var(x, dim = 1, unbiased = False, keepdim = True)
-        mean = torch.mean(x, dim = 1, keepdim = True)
+        self.g = mindspore.Parameter(ops.ones(size = 1, dtype = 1))  # 'torch.ones':没有对应的mindspore参数 'out';; 'torch.ones':没有对应的mindspore参数 'layout';; 'torch.ones':没有对应的mindspore参数 'device';; 'torch.ones':没有对应的mindspore参数 'requires_grad';
+        self.b = mindspore.Parameter(ops.zeros(size = 1, dtype = 1))  # 'torch.zeros':没有对应的mindspore参数 'out';; 'torch.zeros':没有对应的mindspore参数 'layout';; 'torch.zeros':没有对应的mindspore参数 'device';; 'torch.zeros':没有对应的mindspore参数 'requires_grad';
+
+    def forward(self, x):
+        var = ops.var(input = x, dim = 1, keepdim = True)  # 'torch.var':没有对应的mindspore参数 'out';
+        mean = ops.mean(input = x, dim = 1, keepdim = True)
         return (x - mean) / (var + self.eps).sqrt() * self.g + self.b
 
-class FeedForward(nn.Module):
+class FeedForward(nn.Cell):
     def __init__(self, dim, mult = 4, dropout = 0.):
         super().__init__()
-        self.net = nn.Sequential(
+        self.net = nn.SequentialCell(
             LayerNorm(dim),
-            nn.Conv2d(dim, dim * mult, 1),
+            nn.Conv2d(in_channels = dim, out_channels = dim * mult, kernel_size = 1),
             nn.GELU(),
-            nn.Dropout(dropout),
-            nn.Conv2d(dim * mult, dim, 1),
-            nn.Dropout(dropout)
-        )
+            nn.Dropout(p = dropout),
+            nn.Conv2d(in_channels = dim * mult, out_channels = dim, kernel_size = 1),
+            nn.Dropout(p = dropout)
+        )  # 'torch.nn.Conv2d':没有对应的mindspore参数 'device';
     def forward(self, x):
         return self.net(x)
 
-class PatchEmbedding(nn.Module):
+class PatchEmbedding(nn.Cell):
     def __init__(self, *, dim, dim_out, patch_size):
         super().__init__()
         self.dim = dim
         self.dim_out = dim_out
         self.patch_size = patch_size
 
-        self.proj = nn.Sequential(
+        self.proj = nn.SequentialCell(
             LayerNorm(patch_size ** 2 * dim),
-            nn.Conv2d(patch_size ** 2 * dim, dim_out, 1),
+            nn.Conv2d(in_channels = patch_size ** 2 * dim, out_channels = dim_out, kernel_size = 1),
             LayerNorm(dim_out)
-        )
+        )  # 'torch.nn.Conv2d':没有对应的mindspore参数 'device';
 
     def forward(self, fmap):
         p = self.patch_size
         fmap = rearrange(fmap, 'b c (h p1) (w p2) -> b (c p1 p2) h w', p1 = p, p2 = p)
         return self.proj(fmap)
 
-class PEG(nn.Module):
+class PEG(nn.Cell):
     def __init__(self, dim, kernel_size = 3):
         super().__init__()
-        self.proj = Residual(nn.Conv2d(dim, dim, kernel_size = kernel_size, padding = kernel_size // 2, groups = dim, stride = 1))
+        self.proj = Residual(nn.Conv2d(in_channels = dim, out_channels = dim, kernel_size = kernel_size, stride = 1, padding = kernel_size // 2, groups = dim))  # 'torch.nn.Conv2d':没有对应的mindspore参数 'device';
 
     def forward(self, x):
         return self.proj(x)
 
-class LocalAttention(nn.Module):
+class LocalAttention(nn.Cell):
     def __init__(self, dim, heads = 8, dim_head = 64, dropout = 0., patch_size = 7):
         super().__init__()
         inner_dim = dim_head *  heads
@@ -91,13 +90,13 @@         self.scale = dim_head ** -0.5
 
         self.norm = LayerNorm(dim)
-        self.to_q = nn.Conv2d(dim, inner_dim, 1, bias = False)
-        self.to_kv = nn.Conv2d(dim, inner_dim * 2, 1, bias = False)
-
-        self.to_out = nn.Sequential(
-            nn.Conv2d(inner_dim, dim, 1),
-            nn.Dropout(dropout)
-        )
+        self.to_q = nn.Conv2d(in_channels = dim, out_channels = inner_dim, kernel_size = 1, bias = False)  # 'torch.nn.Conv2d':没有对应的mindspore参数 'device';
+        self.to_kv = nn.Conv2d(in_channels = dim, out_channels = inner_dim * 2, kernel_size = 1, bias = False)  # 'torch.nn.Conv2d':没有对应的mindspore参数 'device';
+
+        self.to_out = nn.SequentialCell(
+            nn.Conv2d(in_channels = inner_dim, out_channels = dim, kernel_size = 1),
+            nn.Dropout(p = dropout)
+        )  # 'torch.nn.Conv2d':没有对应的mindspore参数 'device';
 
     def forward(self, fmap):
         fmap = self.norm(fmap)
@@ -111,15 +110,15 @@         q, k, v = (self.to_q(fmap), *self.to_kv(fmap).chunk(2, dim = 1))
         q, k, v = map(lambda t: rearrange(t, 'b (h d) p1 p2 -> (b h) (p1 p2) d', h = h), (q, k, v))
 
-        dots = einsum('b i d, b j d -> b i j', q, k) * self.scale
+        dots = ops.einsum(equation = 'b i d, b j d -> b i j', operands = q) * self.scale
 
         attn = dots.softmax(dim = - 1)
 
-        out = einsum('b i j, b j d -> b i d', attn, v)
+        out = ops.einsum(equation = 'b i j, b j d -> b i d', operands = attn)
         out = rearrange(out, '(b x y h) (p1 p2) d -> b (h d) (x p1) (y p2)', h = h, x = x, y = y, p1 = p, p2 = p)
         return self.to_out(out)
 
-class GlobalAttention(nn.Module):
+class GlobalAttention(nn.Cell):
     def __init__(self, dim, heads = 8, dim_head = 64, dropout = 0., k = 7):
         super().__init__()
         inner_dim = dim_head *  heads
@@ -128,15 +127,15 @@ 
         self.norm = LayerNorm(dim)
 
-        self.to_q = nn.Conv2d(dim, inner_dim, 1, bias = False)
-        self.to_kv = nn.Conv2d(dim, inner_dim * 2, k, stride = k, bias = False)
-
-        self.dropout = nn.Dropout(dropout)
-
-        self.to_out = nn.Sequential(
-            nn.Conv2d(inner_dim, dim, 1),
-            nn.Dropout(dropout)
-        )
+        self.to_q = nn.Conv2d(in_channels = dim, out_channels = inner_dim, kernel_size = 1, bias = False)  # 'torch.nn.Conv2d':没有对应的mindspore参数 'device';
+        self.to_kv = nn.Conv2d(in_channels = dim, out_channels = inner_dim * 2, kernel_size = k, stride = k, bias = False)  # 'torch.nn.Conv2d':没有对应的mindspore参数 'device';
+
+        self.dropout = nn.Dropout(p = dropout)
+
+        self.to_out = nn.SequentialCell(
+            nn.Conv2d(in_channels = inner_dim, out_channels = dim, kernel_size = 1),
+            nn.Dropout(p = dropout)
+        )  # 'torch.nn.Conv2d':没有对应的mindspore参数 'device';
 
     def forward(self, x):
         x = self.norm(x)
@@ -147,21 +146,21 @@ 
         q, k, v = map(lambda t: rearrange(t, 'b (h d) x y -> (b h) (x y) d', h = h), (q, k, v))
 
-        dots = einsum('b i d, b j d -> b i j', q, k) * self.scale
+        dots = ops.einsum(equation = 'b i d, b j d -> b i j', operands = q) * self.scale
 
         attn = dots.softmax(dim = -1)
         attn = self.dropout(attn)
 
-        out = einsum('b i j, b j d -> b i d', attn, v)
+        out = ops.einsum(equation = 'b i j, b j d -> b i d', operands = attn)
         out = rearrange(out, '(b h) (x y) d -> b (h d) x y', h = h, y = y)
         return self.to_out(out)
 
-class Transformer(nn.Module):
+class Transformer(nn.Cell):
     def __init__(self, dim, depth, heads = 8, dim_head = 64, mlp_mult = 4, local_patch_size = 7, global_k = 7, dropout = 0., has_local = True):
         super().__init__()
-        self.layers = nn.ModuleList([])
+        self.layers = nn.CellList([])
         for _ in range(depth):
-            self.layers.append(nn.ModuleList([
+            self.layers.append(nn.CellList([
                 Residual(LocalAttention(dim, heads = heads, dim_head = dim_head, dropout = dropout, patch_size = local_patch_size)) if has_local else nn.Identity(),
                 Residual(FeedForward(dim, mlp_mult, dropout = dropout)) if has_local else nn.Identity(),
                 Residual(GlobalAttention(dim, heads = heads, dim_head = dim_head, dropout = dropout, k = global_k)),
@@ -175,7 +174,7 @@             x = ff2(x)
         return x
 
-class TwinsSVT(nn.Module):
+class TwinsSVT(nn.Cell):
     def __init__(
         self,
         *,
@@ -215,7 +214,7 @@ 
             dim_next = config['emb_dim']
 
-            layers.append(nn.Sequential(
+            layers.append(nn.SequentialCell(
                 PatchEmbedding(dim = dim, dim_out = dim_next, patch_size = config['patch_size']),
                 Transformer(dim = dim_next, depth = 1, local_patch_size = config['local_patch_size'], global_k = config['global_k'], dropout = dropout, has_local = not is_last),
                 PEG(dim = dim_next, kernel_size = peg_kernel_size),
@@ -224,12 +223,12 @@ 
             dim = dim_next
 
-        self.layers = nn.Sequential(
+        self.layers = nn.SequentialCell(
             *layers,
-            nn.AdaptiveAvgPool2d(1),
+            nn.AdaptiveAvgPool2d(output_size = 1),
             Rearrange('... () () -> ...'),
-            nn.Linear(dim, num_classes)
-        )
+            nn.Linear(in_features = dim, out_features = num_classes)
+        )  # 'torch.nn.Linear':没有对应的mindspore参数 'device';
 
     def forward(self, x):
         return self.layers(x)
