--- pytorch+++ mindspore@@ -1,9 +1,8 @@-import torch
 from torch import nn, einsum
-import torch.nn.functional as F
 
 from einops import rearrange, repeat
 from einops.layers.torch import Rearrange
+from mindspore.mint import nn, ops
 
 # helpers
 
@@ -15,40 +14,40 @@ 
 # feedforward
 
-class FeedForward(nn.Module):
+class FeedForward(nn.Cell):
     def __init__(self, dim, hidden_dim, dropout = 0.):
         super().__init__()
         self.net = nn.Sequential(
-            nn.LayerNorm(dim),
-            nn.Linear(dim, hidden_dim),
+            nn.LayerNorm(normalized_shape = dim),
+            nn.Linear(in_features = dim, out_features = hidden_dim),
             nn.GELU(),
-            nn.Dropout(dropout),
-            nn.Linear(hidden_dim, dim),
-            nn.Dropout(dropout)
-        )
+            nn.Dropout(p = dropout),
+            nn.Linear(in_features = hidden_dim, out_features = dim),
+            nn.Dropout(p = dropout)
+        )  # 'torch.nn.LayerNorm':没有对应的mindspore参数 'device';; 'torch.nn.Linear':没有对应的mindspore参数 'device';
     def forward(self, x):
         return self.net(x)
 
 # attention
 
-class Attention(nn.Module):
+class Attention(nn.Cell):
     def __init__(self, dim, heads = 8, dim_head = 64, dropout = 0.):
         super().__init__()
         inner_dim = dim_head *  heads
         self.heads = heads
         self.scale = dim_head ** -0.5
 
-        self.norm = nn.LayerNorm(dim)
+        self.norm = nn.LayerNorm(normalized_shape = dim)  # 'torch.nn.LayerNorm':没有对应的mindspore参数 'device';
         self.attend = nn.Softmax(dim = -1)
-        self.dropout = nn.Dropout(dropout)
-
-        self.to_q = nn.Linear(dim, inner_dim, bias = False)
-        self.to_kv = nn.Linear(dim, inner_dim * 2, bias = False)
+        self.dropout = nn.Dropout(p = dropout)
+
+        self.to_q = nn.Linear(in_features = dim, out_features = inner_dim, bias = False)  # 'torch.nn.Linear':没有对应的mindspore参数 'device';
+        self.to_kv = nn.Linear(in_features = dim, out_features = inner_dim * 2, bias = False)  # 'torch.nn.Linear':没有对应的mindspore参数 'device';
 
         self.to_out = nn.Sequential(
-            nn.Linear(inner_dim, dim),
-            nn.Dropout(dropout)
-        )
+            nn.Linear(in_features = inner_dim, out_features = dim),
+            nn.Dropout(p = dropout)
+        )  # 'torch.nn.Linear':没有对应的mindspore参数 'device';
 
     def forward(self, x, context = None, kv_include_self = False):
         b, n, _, h = *x.shape, self.heads
@@ -56,27 +55,27 @@         context = default(context, x)
 
         if kv_include_self:
-            context = torch.cat((x, context), dim = 1) # cross attention requires CLS token includes itself as key / value
+            context = ops.cat(tensors = (x, context), dim = 1)  # cross attention requires CLS token includes itself as key / value; 'torch.cat':没有对应的mindspore参数 'out';
 
         qkv = (self.to_q(x), *self.to_kv(context).chunk(2, dim = -1))
         q, k, v = map(lambda t: rearrange(t, 'b n (h d) -> b h n d', h = h), qkv)
 
-        dots = einsum('b h i d, b h j d -> b h i j', q, k) * self.scale
+        dots = ops.einsum(equation = 'b h i d, b h j d -> b h i j', operands = q) * self.scale
 
         attn = self.attend(dots)
         attn = self.dropout(attn)
 
-        out = einsum('b h i j, b h j d -> b h i d', attn, v)
+        out = ops.einsum(equation = 'b h i j, b h j d -> b h i d', operands = attn)
         out = rearrange(out, 'b h n d -> b n (h d)')
         return self.to_out(out)
 
 # transformer encoder, for small and large patches
 
-class Transformer(nn.Module):
+class Transformer(nn.Cell):
     def __init__(self, dim, depth, heads, dim_head, mlp_dim, dropout = 0.):
         super().__init__()
         self.layers = nn.ModuleList([])
-        self.norm = nn.LayerNorm(dim)
+        self.norm = nn.LayerNorm(normalized_shape = dim)  # 'torch.nn.LayerNorm':没有对应的mindspore参数 'device';
         for _ in range(depth):
             self.layers.append(nn.ModuleList([
                 Attention(dim, heads = heads, dim_head = dim_head, dropout = dropout),
@@ -91,14 +90,14 @@ 
 # projecting CLS tokens, in the case that small and large patch tokens have different dimensions
 
-class ProjectInOut(nn.Module):
+class ProjectInOut(nn.Cell):
     def __init__(self, dim_in, dim_out, fn):
         super().__init__()
         self.fn = fn
 
         need_projection = dim_in != dim_out
-        self.project_in = nn.Linear(dim_in, dim_out) if need_projection else nn.Identity()
-        self.project_out = nn.Linear(dim_out, dim_in) if need_projection else nn.Identity()
+        self.project_in = nn.Linear(in_features = dim_in, out_features = dim_out) if need_projection else nn.Identity()  # 'torch.nn.Linear':没有对应的mindspore参数 'device';
+        self.project_out = nn.Linear(in_features = dim_out, out_features = dim_in) if need_projection else nn.Identity()  # 'torch.nn.Linear':没有对应的mindspore参数 'device';
 
     def forward(self, x, *args, **kwargs):
         x = self.project_in(x)
@@ -108,7 +107,7 @@ 
 # cross attention transformer
 
-class CrossTransformer(nn.Module):
+class CrossTransformer(nn.Cell):
     def __init__(self, sm_dim, lg_dim, depth, heads, dim_head, dropout):
         super().__init__()
         self.layers = nn.ModuleList([])
@@ -125,13 +124,13 @@             sm_cls = sm_attend_lg(sm_cls, context = lg_patch_tokens, kv_include_self = True) + sm_cls
             lg_cls = lg_attend_sm(lg_cls, context = sm_patch_tokens, kv_include_self = True) + lg_cls
 
-        sm_tokens = torch.cat((sm_cls, sm_patch_tokens), dim = 1)
-        lg_tokens = torch.cat((lg_cls, lg_patch_tokens), dim = 1)
+        sm_tokens = ops.cat(tensors = (sm_cls, sm_patch_tokens), dim = 1)  # 'torch.cat':没有对应的mindspore参数 'out';
+        lg_tokens = ops.cat(tensors = (lg_cls, lg_patch_tokens), dim = 1)  # 'torch.cat':没有对应的mindspore参数 'out';
         return sm_tokens, lg_tokens
 
 # multi-scale encoder
 
-class MultiScaleEncoder(nn.Module):
+class MultiScaleEncoder(nn.Cell):
     def __init__(
         self,
         *,
@@ -163,7 +162,7 @@ 
 # patch-based image to token embedder
 
-class ImageEmbedder(nn.Module):
+class ImageEmbedder(nn.Cell):
     def __init__(
         self,
         *,
@@ -180,28 +179,28 @@ 
         self.to_patch_embedding = nn.Sequential(
             Rearrange('b c (h p1) (w p2) -> b (h w) (p1 p2 c)', p1 = patch_size, p2 = patch_size),
-            nn.LayerNorm(patch_dim),
-            nn.Linear(patch_dim, dim),
-            nn.LayerNorm(dim)
-        )
-
-        self.pos_embedding = nn.Parameter(torch.randn(1, num_patches + 1, dim))
-        self.cls_token = nn.Parameter(torch.randn(1, 1, dim))
-        self.dropout = nn.Dropout(dropout)
+            nn.LayerNorm(normalized_shape = patch_dim),
+            nn.Linear(in_features = patch_dim, out_features = dim),
+            nn.LayerNorm(normalized_shape = dim)
+        )  # 'torch.nn.LayerNorm':没有对应的mindspore参数 'device';; 'torch.nn.Linear':没有对应的mindspore参数 'device';
+
+        self.pos_embedding = nn.Parameter(ops.randn(size = 1, generator = num_patches + 1))  # 'torch.randn':没有对应的mindspore参数 'out';; 'torch.randn':没有对应的mindspore参数 'layout';; 'torch.randn':没有对应的mindspore参数 'device';; 'torch.randn':没有对应的mindspore参数 'requires_grad';; 'torch.randn':没有对应的mindspore参数 'pin_memory';
+        self.cls_token = nn.Parameter(ops.randn(size = 1, generator = 1))  # 'torch.randn':没有对应的mindspore参数 'out';; 'torch.randn':没有对应的mindspore参数 'layout';; 'torch.randn':没有对应的mindspore参数 'device';; 'torch.randn':没有对应的mindspore参数 'requires_grad';; 'torch.randn':没有对应的mindspore参数 'pin_memory';
+        self.dropout = nn.Dropout(p = dropout)
 
     def forward(self, img):
         x = self.to_patch_embedding(img)
         b, n, _ = x.shape
 
         cls_tokens = repeat(self.cls_token, '() n d -> b n d', b = b)
-        x = torch.cat((cls_tokens, x), dim=1)
+        x = ops.cat(tensors = (cls_tokens, x), dim = 1)  # 'torch.cat':没有对应的mindspore参数 'out';
         x += self.pos_embedding[:, :(n + 1)]
 
         return self.dropout(x)
 
 # cross ViT class
 
-class CrossViT(nn.Module):
+class CrossViT(nn.Cell):
     def __init__(
         self,
         *,
@@ -253,8 +252,8 @@             dropout = dropout
         )
 
-        self.sm_mlp_head = nn.Sequential(nn.LayerNorm(sm_dim), nn.Linear(sm_dim, num_classes))
-        self.lg_mlp_head = nn.Sequential(nn.LayerNorm(lg_dim), nn.Linear(lg_dim, num_classes))
+        self.sm_mlp_head = nn.Sequential(nn.LayerNorm(normalized_shape = sm_dim), nn.Linear(in_features = sm_dim, out_features = num_classes))  # 'torch.nn.LayerNorm':没有对应的mindspore参数 'device';; 'torch.nn.Linear':没有对应的mindspore参数 'device';
+        self.lg_mlp_head = nn.Sequential(nn.LayerNorm(normalized_shape = lg_dim), nn.Linear(in_features = lg_dim, out_features = num_classes))  # 'torch.nn.LayerNorm':没有对应的mindspore参数 'device';; 'torch.nn.Linear':没有对应的mindspore参数 'device';
 
     def forward(self, img):
         sm_tokens = self.sm_image_embedder(img)
