--- pytorch+++ mindspore@@ -1,3 +1,4 @@+from mindspore.mint import nn, ops
 from math import sqrt, pi, log
 
 import torch
@@ -14,21 +15,21 @@ def rotate_every_two(x):
     x = rearrange(x, '... (d j) -> ... d j', j = 2)
     x1, x2 = x.unbind(dim = -1)
-    x = torch.stack((-x2, x1), dim = -1)
+    x = ops.stack(tensors = (-x2, x1), dim = -1)  # 'torch.stack':没有对应的mindspore参数 'out';
     return rearrange(x, '... d j -> ... (d j)')
 
 class AxialRotaryEmbedding(nn.Module):
     def __init__(self, dim, max_freq = 10):
         super().__init__()
         self.dim = dim
-        scales = torch.linspace(1., max_freq / 2, self.dim // 4)
+        scales = ops.linspace(start = 1., end = max_freq / 2, steps = self.dim // 4)  # 'torch.linspace':没有对应的mindspore参数 'out';; 'torch.linspace':没有对应的mindspore参数 'layout';; 'torch.linspace':没有对应的mindspore参数 'device';; 'torch.linspace':没有对应的mindspore参数 'requires_grad';
         self.register_buffer('scales', scales)
 
     @autocast('cuda', enabled = False)
     def forward(self, x):
         device, dtype, n = x.device, x.dtype, int(sqrt(x.shape[-2]))
 
-        seq = torch.linspace(-1., 1., steps = n, device = device)
+        seq = ops.linspace(start = -1., end = 1., steps = n)  # 'torch.linspace':没有对应的mindspore参数 'out';; 'torch.linspace':没有对应的mindspore参数 'layout';; 'torch.linspace':没有对应的mindspore参数 'device';; 'torch.linspace':没有对应的mindspore参数 'requires_grad';
         seq = seq.unsqueeze(-1)
 
         scales = self.scales[(*((None,) * (len(seq.shape) - 1)), Ellipsis)]
@@ -39,8 +40,8 @@         x_sinu = repeat(seq, 'i d -> i j d', j = n)
         y_sinu = repeat(seq, 'j d -> i j d', i = n)
 
-        sin = torch.cat((x_sinu.sin(), y_sinu.sin()), dim = -1)
-        cos = torch.cat((x_sinu.cos(), y_sinu.cos()), dim = -1)
+        sin = ops.cat(tensors = (x_sinu.sin(), y_sinu.sin()), dim = -1)  # 'torch.cat':没有对应的mindspore参数 'out';
+        cos = ops.cat(tensors = (x_sinu.cos(), y_sinu.cos()), dim = -1)  # 'torch.cat':没有对应的mindspore参数 'out';
 
         sin, cos = map(lambda t: rearrange(t, 'i j d -> (i j) d'), (sin, cos))
         sin, cos = map(lambda t: repeat(t, 'n d -> () n (d j)', j = 2), (sin, cos))
@@ -50,9 +51,9 @@     def __init__(self, dim_in, dim_out, kernel_size, padding, stride = 1, bias = True):
         super().__init__()
         self.net = nn.Sequential(
-            nn.Conv2d(dim_in, dim_in, kernel_size = kernel_size, padding = padding, groups = dim_in, stride = stride, bias = bias),
-            nn.Conv2d(dim_in, dim_out, kernel_size = 1, bias = bias)
-        )
+            nn.Conv2d(in_channels = dim_in, out_channels = dim_in, kernel_size = kernel_size, stride = stride, padding = padding, groups = dim_in, bias = bias),
+            nn.Conv2d(in_channels = dim_in, out_channels = dim_out, kernel_size = 1, bias = bias)
+        )  # 'torch.nn.Conv2d':没有对应的mindspore参数 'device';
     def forward(self, x):
         return self.net(x)
 
@@ -62,7 +63,7 @@     def __init__(self, dim_in, dim_out, kernel, bias = False):
         super().__init__()
         self.conv = DepthWiseConv2d(dim_in, dim_out, kernel, padding = kernel // 2, bias = False)
-        self.cls_proj = nn.Linear(dim_in, dim_out) if dim_in != dim_out else nn.Identity()
+        self.cls_proj = nn.Linear(in_features = dim_in, out_features = dim_out) if dim_in != dim_out else nn.Identity()  # 'torch.nn.Linear':没有对应的mindspore参数 'device';
 
     def forward(self, x, fmap_dims):
         cls_token, x = x[:, :1], x[:, 1:]
@@ -70,24 +71,24 @@         x = self.conv(x)
         x = rearrange(x, 'b d h w -> b (h w) d')
         cls_token = self.cls_proj(cls_token)
-        return torch.cat((cls_token, x), dim = 1)
+        return ops.cat(tensors = (cls_token, x), dim = 1)  # 'torch.cat':没有对应的mindspore参数 'out';
 
 class GEGLU(nn.Module):
     def forward(self, x):
         x, gates = x.chunk(2, dim = -1)
-        return F.gelu(gates) * x
+        return nn.functional.gelu(input = gates) * x  # 'torch.nn.functional.gelu':没有对应的mindspore参数 'dim';
 
 class FeedForward(nn.Module):
     def __init__(self, dim, hidden_dim, dropout = 0., use_glu = True):
         super().__init__()
         self.net = nn.Sequential(
-            nn.LayerNorm(dim),
-            nn.Linear(dim, hidden_dim * 2 if use_glu else hidden_dim),
+            nn.LayerNorm(normalized_shape = dim),
+            nn.Linear(in_features = dim, out_features = hidden_dim * 2 if use_glu else hidden_dim),
             GEGLU() if use_glu else nn.GELU(),
-            nn.Dropout(dropout),
-            nn.Linear(hidden_dim, dim),
-            nn.Dropout(dropout)
-        )
+            nn.Dropout(p = dropout),
+            nn.Linear(in_features = hidden_dim, out_features = dim),
+            nn.Dropout(p = dropout)
+        )  # 'torch.nn.LayerNorm':没有对应的mindspore参数 'device';; 'torch.nn.Linear':没有对应的mindspore参数 'device';
     def forward(self, x):
         return self.net(x)
 
@@ -99,20 +100,20 @@         self.heads = heads
         self.scale = dim_head ** -0.5
 
-        self.norm = nn.LayerNorm(dim)
+        self.norm = nn.LayerNorm(normalized_shape = dim)  # 'torch.nn.LayerNorm':没有对应的mindspore参数 'device';
         self.attend = nn.Softmax(dim = -1)
-        self.dropout = nn.Dropout(dropout)
+        self.dropout = nn.Dropout(p = dropout)
 
         self.use_ds_conv = use_ds_conv
 
-        self.to_q = SpatialConv(dim, inner_dim, conv_query_kernel, bias = False) if use_ds_conv else nn.Linear(dim, inner_dim, bias = False)
-
-        self.to_kv = nn.Linear(dim, inner_dim * 2, bias = False)
+        self.to_q = SpatialConv(dim, inner_dim, conv_query_kernel, bias = False) if use_ds_conv else nn.Linear(in_features = dim, out_features = inner_dim, bias = False)  # 'torch.nn.Linear':没有对应的mindspore参数 'device';
+
+        self.to_kv = nn.Linear(in_features = dim, out_features = inner_dim * 2, bias = False)  # 'torch.nn.Linear':没有对应的mindspore参数 'device';
 
         self.to_out = nn.Sequential(
-            nn.Linear(inner_dim, dim),
-            nn.Dropout(dropout)
-        )
+            nn.Linear(in_features = inner_dim, out_features = dim),
+            nn.Dropout(p = dropout)
+        )  # 'torch.nn.Linear':没有对应的mindspore参数 'device';
 
     def forward(self, x, pos_emb, fmap_dims):
         b, n, _, h = *x.shape, self.heads
@@ -139,19 +140,19 @@ 
             (q, q_pass), (k, k_pass) = map(lambda t: (t[..., :dim_rotary], t[..., dim_rotary:]), (q, k))
             q, k = map(lambda t: (t * cos) + (rotate_every_two(t) * sin), (q, k))
-            q, k = map(lambda t: torch.cat(t, dim = -1), ((q, q_pass), (k, k_pass)))
+            q, k = map(lambda t: ops.cat(tensors = t, dim = -1), ((q, q_pass), (k, k_pass)))  # 'torch.cat':没有对应的mindspore参数 'out';
 
             # concat back the CLS tokens
 
-            q = torch.cat((q_cls, q), dim = 1)
-            k = torch.cat((k_cls, k), dim = 1)
-
-        dots = einsum('b i d, b j d -> b i j', q, k) * self.scale
+            q = ops.cat(tensors = (q_cls, q), dim = 1)  # 'torch.cat':没有对应的mindspore参数 'out';
+            k = ops.cat(tensors = (k_cls, k), dim = 1)  # 'torch.cat':没有对应的mindspore参数 'out';
+
+        dots = ops.einsum(equation = 'b i d, b j d -> b i j', operands = q) * self.scale
 
         attn = self.attend(dots)
         attn = self.dropout(attn)
 
-        out = einsum('b i j, b j d -> b i d', attn, v)
+        out = ops.einsum(equation = 'b i j, b j d -> b i d', operands = attn)
         out = rearrange(out, '(b h) n d -> b n (h d)', h = h)
         return self.to_out(out)
 
@@ -185,16 +186,16 @@         self.patch_size = patch_size
         self.to_patch_embedding = nn.Sequential(
             Rearrange('b c (h p1) (w p2) -> b (h w) (p1 p2 c)', p1 = patch_size, p2 = patch_size),
-            nn.Linear(patch_dim, dim),
-        )
-
-        self.cls_token = nn.Parameter(torch.randn(1, 1, dim))
+            nn.Linear(in_features = patch_dim, out_features = dim),
+        )  # 'torch.nn.Linear':没有对应的mindspore参数 'device';
+
+        self.cls_token = nn.Parameter(ops.randn(size = 1, generator = 1))  # 'torch.randn':没有对应的mindspore参数 'out';; 'torch.randn':没有对应的mindspore参数 'layout';; 'torch.randn':没有对应的mindspore参数 'device';; 'torch.randn':没有对应的mindspore参数 'requires_grad';; 'torch.randn':没有对应的mindspore参数 'pin_memory';
         self.transformer = Transformer(dim, depth, heads, dim_head, mlp_dim, image_size, dropout, use_rotary, use_ds_conv, use_glu)
 
         self.mlp_head = nn.Sequential(
-            nn.LayerNorm(dim),
-            nn.Linear(dim, num_classes)
-        )
+            nn.LayerNorm(normalized_shape = dim),
+            nn.Linear(in_features = dim, out_features = num_classes)
+        )  # 'torch.nn.LayerNorm':没有对应的mindspore参数 'device';; 'torch.nn.Linear':没有对应的mindspore参数 'device';
 
     def forward(self, img):
         b, _, h, w, p = *img.shape, self.patch_size
@@ -203,7 +204,7 @@         n = x.shape[1]
 
         cls_tokens = repeat(self.cls_token, '() n d -> b n d', b = b)
-        x = torch.cat((cls_tokens, x), dim=1)
+        x = ops.cat(tensors = (cls_tokens, x), dim = 1)  # 'torch.cat':没有对应的mindspore参数 'out';
 
         fmap_dims = {'h': h // p, 'w': w // p}
         x = self.transformer(x, fmap_dims = fmap_dims)
