--- pytorch+++ mindspore@@ -1,4 +1,8 @@-import torch
+import mindspore as ms
+import mindspore.nn as msnn
+import mindspore.ops as msops
+import mindspore.mint as mint
+from mindspore.mint import nn, ops
 from torch import nn
 
 from einops import rearrange, repeat, pack, unpack
@@ -6,10 +10,10 @@ 
 # classes
 
-class FeedForward(nn.Module):
+class FeedForward(msnn.Cell):
     def __init__(self, dim, hidden_dim, dropout = 0.):
         super().__init__()
-        self.net = nn.Sequential(
+        self.net = msnn.SequentialCell(
             nn.LayerNorm(dim),
             nn.Linear(dim, hidden_dim),
             nn.GELU(),
@@ -20,7 +24,7 @@     def forward(self, x):
         return self.net(x)
 
-class Attention(nn.Module):
+class Attention(msnn.Cell):
     def __init__(self, dim, heads = 8, dim_head = 64, dropout = 0.):
         super().__init__()
         inner_dim = dim_head *  heads
@@ -35,31 +39,31 @@ 
         self.to_qkv = nn.Linear(dim, inner_dim * 3, bias = False)
 
-        self.to_out = nn.Sequential(
+        self.to_out = msnn.SequentialCell(
             nn.Linear(inner_dim, dim),
             nn.Dropout(dropout)
-        ) if project_out else nn.Identity()
+        ) if project_out else msnn.Identity()
 
     def forward(self, x):
         x = self.norm(x)
         qkv = self.to_qkv(x).chunk(3, dim = -1)
         q, k, v = map(lambda t: rearrange(t, 'b n (h d) -> b h n d', h = self.heads), qkv)
 
-        dots = torch.matmul(q, k.transpose(-1, -2)) * self.scale
+        dots = mint.matmul(q, k.transpose(-1, -2)) * self.scale
 
         attn = self.attend(dots)
         attn = self.dropout(attn)
 
-        out = torch.matmul(attn, v)
+        out = mint.matmul(attn, v)
         out = rearrange(out, 'b h n d -> b n (h d)')
         return self.to_out(out)
 
-class Transformer(nn.Module):
+class Transformer(msnn.Cell):
     def __init__(self, dim, depth, heads, dim_head, mlp_dim, dropout = 0.):
         super().__init__()
-        self.layers = nn.ModuleList([])
+        self.layers = msnn.CellList([])
         for _ in range(depth):
-            self.layers.append(nn.ModuleList([
+            self.layers.append(msnn.CellList([
                 Attention(dim, heads = heads, dim_head = dim_head, dropout = dropout),
                 FeedForward(dim, mlp_dim, dropout = dropout)
             ]))
@@ -69,7 +73,7 @@             x = ff(x) + x
         return x
 
-class ViT(nn.Module):
+class ViT(msnn.Cell):
     def __init__(self, *, seq_len, patch_size, num_classes, dim, depth, heads, mlp_dim, channels = 3, dim_head = 64, dropout = 0., emb_dropout = 0.):
         super().__init__()
         assert (seq_len % patch_size) == 0
@@ -77,20 +81,20 @@         num_patches = seq_len // patch_size
         patch_dim = channels * patch_size
 
-        self.to_patch_embedding = nn.Sequential(
+        self.to_patch_embedding = msnn.SequentialCell(
             Rearrange('b c (n p) -> b n (p c)', p = patch_size),
             nn.LayerNorm(patch_dim),
             nn.Linear(patch_dim, dim),
             nn.LayerNorm(dim),
         )
 
-        self.pos_embedding = nn.Parameter(torch.randn(1, num_patches + 1, dim))
-        self.cls_token = nn.Parameter(torch.randn(dim))
+        self.pos_embedding = ms.Parameter(mint.randn(1, num_patches + 1))  # 'torch.randn':没有对应的mindspore参数 'out' (position 2);
+        self.cls_token = ms.Parameter(mint.randn(dim))
         self.dropout = nn.Dropout(emb_dropout)
 
         self.transformer = Transformer(dim, depth, heads, dim_head, mlp_dim, dropout)
 
-        self.mlp_head = nn.Sequential(
+        self.mlp_head = msnn.SequentialCell(
             nn.LayerNorm(dim),
             nn.Linear(dim, num_classes)
         )
@@ -126,5 +130,5 @@         emb_dropout = 0.1
     )
 
-    time_series = torch.randn(4, 3, 256)
+    time_series = mint.randn(4, 3)  # 'torch.randn':没有对应的mindspore参数 'out' (position 2);
     logits = v(time_series) # (4, 1000)
