--- pytorch+++ mindspore@@ -3,11 +3,10 @@ 
 import torch
 from torch import nn, stack, tensor
-import torch.nn.functional as F
-from torch.nn import Module, ModuleList
 
 from einops import rearrange, repeat, reduce, einsum, pack, unpack
 from einops.layers.torch import Rearrange
+from mindspore.mint import nn, ops
 
 # helpers
 
@@ -22,7 +21,7 @@ 
 # decorr loss
 
-class DecorrelationLoss(Module):
+class DecorrelationLoss(nn.Cell):
     def __init__(
         self,
         sample_frac = 1.,
@@ -52,16 +51,16 @@ 
             tokens, packed_shape = pack([tokens], '* n d e')
 
-            indices = torch.randn(tokens.shape[:2]).argsort(dim = -1)[..., :num_sampled, :]
-
-            batch_arange = torch.arange(tokens.shape[0], device = tokens.device)
+            indices = ops.randn(size = tokens.shape[:2]).argsort(dim = -1)[..., :num_sampled, :]  # 'torch.randn':没有对应的mindspore参数 'out';; 'torch.randn':没有对应的mindspore参数 'layout';; 'torch.randn':没有对应的mindspore参数 'device';; 'torch.randn':没有对应的mindspore参数 'requires_grad';; 'torch.randn':没有对应的mindspore参数 'pin_memory';
+
+            batch_arange = ops.arange(start = tokens.shape[0])  # 'torch.arange':没有对应的mindspore参数 'out';; 'torch.arange':没有对应的mindspore参数 'layout';; 'torch.arange':没有对应的mindspore参数 'device';; 'torch.arange':没有对应的mindspore参数 'requires_grad';
             batch_arange = rearrange(batch_arange, 'b -> b 1')
 
             tokens = tokens[batch_arange, indices]
             tokens, = unpack(tokens, packed_shape, '* n d e')
 
         dist = einsum(tokens, tokens, '... n d, ... n e -> ... d e') / tokens.shape[-2]
-        eye = torch.eye(dim, device = device)
+        eye = ops.eye(n = dim)  # 'torch.eye':没有对应的mindspore参数 'out';; 'torch.eye':没有对应的mindspore参数 'layout';; 'torch.eye':没有对应的mindspore参数 'device';; 'torch.eye':没有对应的mindspore参数 'requires_grad';
 
         loss = dist.pow(2) * (1. - eye) / ((dim - 1) * dim)
 
@@ -70,42 +69,42 @@ 
 # classes
 
-class FeedForward(Module):
+class FeedForward(nn.Cell):
     def __init__(self, dim, hidden_dim, dropout = 0.):
         super().__init__()
-        self.norm = nn.LayerNorm(dim)
-
-        self.net = nn.Sequential(
-            nn.Linear(dim, hidden_dim),
+        self.norm = nn.LayerNorm(normalized_shape = dim)  # 'torch.nn.LayerNorm':没有对应的mindspore参数 'device';
+
+        self.net = nn.SequentialCell(
+            nn.Linear(in_features = dim, out_features = hidden_dim),
             nn.GELU(),
-            nn.Dropout(dropout),
-            nn.Linear(hidden_dim, dim),
-            nn.Dropout(dropout)
-        )
+            nn.Dropout(p = dropout),
+            nn.Linear(in_features = hidden_dim, out_features = dim),
+            nn.Dropout(p = dropout)
+        )  # 'torch.nn.Linear':没有对应的mindspore参数 'device';
 
     def forward(self, x):
         normed = self.norm(x)
         return self.net(x), normed
 
-class Attention(Module):
+class Attention(nn.Cell):
     def __init__(self, dim, heads = 8, dim_head = 64, dropout = 0.):
         super().__init__()
         inner_dim = dim_head *  heads
         project_out = not (heads == 1 and dim_head == dim)
 
-        self.norm = nn.LayerNorm(dim)
+        self.norm = nn.LayerNorm(normalized_shape = dim)  # 'torch.nn.LayerNorm':没有对应的mindspore参数 'device';
         self.heads = heads
         self.scale = dim_head ** -0.5
 
         self.attend = nn.Softmax(dim = -1)
-        self.dropout = nn.Dropout(dropout)
-
-        self.to_qkv = nn.Linear(dim, inner_dim * 3, bias = False)
-
-        self.to_out = nn.Sequential(
-            nn.Linear(inner_dim, dim),
-            nn.Dropout(dropout)
-        ) if project_out else nn.Identity()
+        self.dropout = nn.Dropout(p = dropout)
+
+        self.to_qkv = nn.Linear(in_features = dim, out_features = inner_dim * 3, bias = False)  # 'torch.nn.Linear':没有对应的mindspore参数 'device';
+
+        self.to_out = nn.SequentialCell(
+            nn.Linear(in_features = inner_dim, out_features = dim),
+            nn.Dropout(p = dropout)
+        ) if project_out else nn.Identity()  # 'torch.nn.Linear':没有对应的mindspore参数 'device';
 
     def forward(self, x):
         normed = self.norm(x)
@@ -113,24 +112,24 @@         qkv = self.to_qkv(normed).chunk(3, dim = -1)
         q, k, v = map(lambda t: rearrange(t, 'b n (h d) -> b h n d', h = self.heads), qkv)
 
-        dots = torch.matmul(q, k.transpose(-1, -2)) * self.scale
+        dots = ops.matmul(input = q, other = k.transpose(-1, -2)) * self.scale  # 'torch.matmul':没有对应的mindspore参数 'out';
 
         attn = self.attend(dots)
         attn = self.dropout(attn)
 
-        out = torch.matmul(attn, v)
+        out = ops.matmul(input = attn, other = v)  # 'torch.matmul':没有对应的mindspore参数 'out';
         out = rearrange(out, 'b h n d -> b n (h d)')
 
         return self.to_out(out), normed
 
-class Transformer(Module):
+class Transformer(nn.Cell):
     def __init__(self, dim, depth, heads, dim_head, mlp_dim, dropout = 0.):
         super().__init__()
-        self.norm = nn.LayerNorm(dim)
-        self.layers = ModuleList([])
+        self.norm = nn.LayerNorm(normalized_shape = dim)  # 'torch.nn.LayerNorm':没有对应的mindspore参数 'device';
+        self.layers = nn.CellList([])
 
         for _ in range(depth):
-            self.layers.append(ModuleList([
+            self.layers.append(nn.CellList([
                 Attention(dim, heads = heads, dim_head = dim_head, dropout = dropout),
                 FeedForward(dim, mlp_dim, dropout = dropout)
             ]))
@@ -149,9 +148,9 @@             normed_inputs.append(attn_normed_inp)
             normed_inputs.append(ff_normed_inp)
 
-        return self.norm(x), stack(normed_inputs)
-
-class ViT(Module):
+        return self.norm(x), ops.stack(tensors = normed_inputs)  # 'torch.stack':没有对应的mindspore参数 'out';
+
+class ViT(nn.Cell):
     def __init__(self, *, image_size, patch_size, num_classes, dim, depth, heads, mlp_dim, pool = 'cls', channels = 3, dim_head = 64, dropout = 0., emb_dropout = 0., decorr_sample_frac = 1.):
         super().__init__()
         image_height, image_width = pair(image_size)
@@ -163,23 +162,23 @@         patch_dim = channels * patch_height * patch_width
         assert pool in {'cls', 'mean'}, 'pool type must be either cls (cls token) or mean (mean pooling)'
 
-        self.to_patch_embedding = nn.Sequential(
+        self.to_patch_embedding = nn.SequentialCell(
             Rearrange('b c (h p1) (w p2) -> b (h w) (p1 p2 c)', p1 = patch_height, p2 = patch_width),
-            nn.LayerNorm(patch_dim),
-            nn.Linear(patch_dim, dim),
-            nn.LayerNorm(dim),
-        )
-
-        self.pos_embedding = nn.Parameter(torch.randn(1, num_patches + 1, dim))
-        self.cls_token = nn.Parameter(torch.randn(1, 1, dim))
-        self.dropout = nn.Dropout(emb_dropout)
+            nn.LayerNorm(normalized_shape = patch_dim),
+            nn.Linear(in_features = patch_dim, out_features = dim),
+            nn.LayerNorm(normalized_shape = dim),
+        )  # 'torch.nn.LayerNorm':没有对应的mindspore参数 'device';; 'torch.nn.Linear':没有对应的mindspore参数 'device';
+
+        self.pos_embedding = mindspore.Parameter(ops.randn(size = 1, generator = num_patches + 1))  # 'torch.randn':没有对应的mindspore参数 'out';; 'torch.randn':没有对应的mindspore参数 'layout';; 'torch.randn':没有对应的mindspore参数 'device';; 'torch.randn':没有对应的mindspore参数 'requires_grad';; 'torch.randn':没有对应的mindspore参数 'pin_memory';
+        self.cls_token = mindspore.Parameter(ops.randn(size = 1, generator = 1))  # 'torch.randn':没有对应的mindspore参数 'out';; 'torch.randn':没有对应的mindspore参数 'layout';; 'torch.randn':没有对应的mindspore参数 'device';; 'torch.randn':没有对应的mindspore参数 'requires_grad';; 'torch.randn':没有对应的mindspore参数 'pin_memory';
+        self.dropout = nn.Dropout(p = emb_dropout)
 
         self.transformer = Transformer(dim, depth, heads, dim_head, mlp_dim, dropout)
 
         self.pool = pool
         self.to_latent = nn.Identity()
 
-        self.mlp_head = nn.Linear(dim, num_classes)
+        self.mlp_head = nn.Linear(in_features = dim, out_features = num_classes)  # 'torch.nn.Linear':没有对应的mindspore参数 'device';
 
         # decorrelation loss related
 
@@ -201,7 +200,7 @@         b, n, _ = x.shape
 
         cls_tokens = repeat(self.cls_token, '1 1 d -> b 1 d', b = b)
-        x = torch.cat((cls_tokens, x), dim=1)
+        x = ops.cat(tensors = (cls_tokens, x), dim = 1)  # 'torch.cat':没有对应的mindspore参数 'out';
         x += self.pos_embedding[:, :(n + 1)]
         x = self.dropout(x)
 
@@ -224,7 +223,7 @@ if __name__ == '__main__':
     decorr_loss = DecorrelationLoss(0.1)
 
-    hiddens = torch.randn(6, 2, 512, 256)
+    hiddens = ops.randn(size = 6, generator = 2, dtype = 256)  # 'torch.randn':没有对应的mindspore参数 'out';; 'torch.randn':没有对应的mindspore参数 'layout';; 'torch.randn':没有对应的mindspore参数 'device';; 'torch.randn':没有对应的mindspore参数 'requires_grad';; 'torch.randn':没有对应的mindspore参数 'pin_memory';
 
     decorr_loss(hiddens)
     decorr_loss(hiddens[0])
