--- pytorch+++ mindspore@@ -1,3 +1,8 @@+import mindspore as ms
+import mindspore.nn as msnn
+import mindspore.ops as msops
+import mindspore.mint as mint
+from mindspore.mint import nn, ops
 """ A dataset reader that reads tarfile based datasets
 
 This reader can extract image samples from:
@@ -123,7 +128,7 @@     tarfiles = []
 
     def _label_from_paths(*path, leaf_only=True):
-        path = os.path.join(*path).strip(os.path.sep)
+        path = os.path.join(*path).strip(os.path.sep)  # 存在 *args/**kwargs，未转换，需手动确认参数映射;
         return path.split(os.path.sep)[-1] if leaf_only else path.replace(os.path.sep, '_')
 
     def _add_samples(info, fn):
@@ -162,7 +167,7 @@     samples_and_targets = [(s, class_name_to_idx[l]) for s, l in zip(samples, labels) if l in class_name_to_idx]
     if sort:
         samples_and_targets = sorted(samples_and_targets, key=lambda k: natural_key(k[0][0].path))
-    samples, targets = zip(*samples_and_targets)
+    samples, targets = zip(*samples_and_targets)  # 存在 *args/**kwargs，未转换，需手动确认参数映射;
     samples = np.array(samples)
     targets = np.array(targets)
     _logger.info(f'Finished processing {len(samples)} samples across {len(tarfiles)} tar files.')
