--- pytorch+++ mindspore@@ -1,3 +1,8 @@+import mindspore as ms
+import mindspore.nn as msnn
+import mindspore.ops as msops
+import mindspore.mint as mint
+from mindspore.mint import nn, ops
 """ Dataset reader that wraps TFDS datasets
 
 Wraps many (most?) TFDS image-classification datasets
@@ -11,8 +16,8 @@ import sys
 from typing import Optional
 
-import torch
-import torch.distributed as dist
+# import torch
+# import torch.distributed as dist
 from PIL import Image
 
 try:
@@ -163,9 +168,11 @@         # Distributed world state
         self.dist_rank = 0
         self.dist_num_replicas = 1
-        if dist.is_available() and dist.is_initialized() and dist.get_world_size() > 1:
-            self.dist_rank = dist.get_rank()
-            self.dist_num_replicas = dist.get_world_size()
+        # 'torch.distributed.is_available' 未在映射表(api_mapping_out_excel.json)中找到，需手动确认;
+        # 'torch.distributed.is_initialized' 未在映射表(api_mapping_out_excel.json)中找到，需手动确认;
+        if dist.is_available() and dist.is_initialized() and mint.distributed.get_world_size() > 1:
+            self.dist_rank = mint.distributed.get_rank()
+            self.dist_num_replicas = mint.distributed.get_world_size()
 
         # Attributes that are updated in _lazy_init, including the tf.data pipeline itself
         self.global_num_workers = 1
@@ -204,7 +211,7 @@         instances once it has been initialized. Do not call any dataset methods that can call _lazy_init
         before it is passed to dataloader.
         """
-        worker_info = torch.utils.data.get_worker_info()
+        worker_info = torch.utils.data.get_worker_info()  # 'torch.utils.data.get_worker_info' 未在映射表(api_mapping_out_excel.json)中找到，需手动确认;
 
         # setup input context to split dataset across distributed processes
         num_workers = 1
