--- pytorch+++ mindspore@@ -1,3 +1,8 @@+import mindspore as ms
+import mindspore.nn as msnn
+import mindspore.ops as msops
+import mindspore.mint as mint
+from mindspore.mint import nn, ops
 import os
 import zipfile
 
@@ -10,12 +15,12 @@ #       Ref. PyTorch #18758
 #         https://github.com/pytorch/pytorch/pull/18758/commits
 try:
-    from torch.utils.model_zoo import _download_url_to_file
+    # from torch.utils.model_zoo import _download_url_to_file
 except ImportError:
     try:
-        from torch.hub import download_url_to_file as _download_url_to_file
+        # from torch.hub import download_url_to_file as _download_url_to_file
     except ImportError:
-        from torch.hub import _download_url_to_file
+        # from torch.hub import _download_url_to_file
 
 
 def unzip(source_filename, dest_dir):
@@ -24,5 +29,5 @@ 
 
 if __name__ == '__main__':
-    _download_url_to_file('https://www.dropbox.com/s/lrvwfehqdcxoza8/saved_models.zip?dl=1', 'saved_models.zip', None, True)
+    _download_url_to_file('https://www.dropbox.com/s/lrvwfehqdcxoza8/saved_models.zip?dl=1', 'saved_models.zip', None, True)  # 'torch.hub._download_url_to_file' 未在映射表(api_mapping_out_excel.json)中找到，需手动确认;
     unzip('saved_models.zip', '.')
