--- pytorch+++ mindspore@@ -1,3 +1,8 @@+import mindspore as ms
+import mindspore.nn as msnn
+import mindspore.ops as msops
+import mindspore.mint as mint
+from mindspore.mint import nn, ops
 import dataclasses
 import logging
 import os
@@ -5,8 +10,8 @@ from pathlib import Path
 from typing import Any, Callable, Dict, List, Optional, Tuple, Type, TypeVar, Union
 
-from torch import nn as nn
-from torch.hub import load_state_dict_from_url
+# from torch import nn as nn
+# from torch.hub import load_state_dict_from_url
 
 from timm.models._features import FeatureListNet, FeatureDictNet, FeatureHookNet, FeatureGetterNet
 from timm.models._features_fx import FeatureGraphNet
@@ -100,6 +105,7 @@     _CHECK_HASH = enable
 
 
+# 类型标注 'torch.nn.Module' 未在映射表(api_mapping_out_excel.json)中找到，需手动确认;
 def load_custom_pretrained(
         model: nn.Module,
         pretrained_cfg: Optional[Dict[str, Any]] = None,
@@ -149,6 +155,7 @@         _logger.warning("Valid function to load pretrained weights is not available, using random initialization.")
 
 
+# 类型标注 'torch.nn.Module' 未在映射表(api_mapping_out_excel.json)中找到，需手动确认;
 def load_pretrained(
         model: nn.Module,
         pretrained_cfg: Optional[Dict[str, Any]] = None,
@@ -204,7 +211,7 @@                     check_hash=_CHECK_HASH,
                     weights_only=True,
                     model_dir=cache_dir,
-                )
+                )  # 'torch.hub.load_state_dict_from_url' 未在映射表(api_mapping_out_excel.json)中找到，需手动确认;
             except TypeError:
                 state_dict = load_state_dict_from_url(
                     pretrained_loc,
@@ -212,7 +219,7 @@                     progress=_DOWNLOAD_PROGRESS,
                     check_hash=_CHECK_HASH,
                     model_dir=cache_dir,
-                )
+                )  # 'torch.hub.load_state_dict_from_url' 未在映射表(api_mapping_out_excel.json)中找到，需手动确认;
     elif load_from == 'hf-hub':
         _logger.info(f'Loading pretrained weights from Hugging Face hub ({pretrained_loc})')
         if isinstance(pretrained_loc, (list, tuple)):
