--- pytorch+++ mindspore@@ -1,9 +1,12 @@+import mindspore as ms
+import mindspore.nn as msnn
+import mindspore.ops as msops
+import mindspore.mint as mint
+from mindspore.mint import nn, ops
 import os
 from pathlib import Path
 from typing import Any, Dict, Optional, Tuple, Union
 from urllib.parse import urlsplit
-
-from torch import nn
 
 from timm.layers import set_layer_config
 from ._helpers import load_checkpoint
@@ -52,7 +55,7 @@         exportable: Optional[bool] = None,
         no_jit: Optional[bool] = None,
         **kwargs: Any,
-) -> nn.Module:
+) -> msnn.Cell:
     """Create a model.
 
     Lookup model's entrypoint function and pass relevant args to create a new model.
@@ -141,7 +144,7 @@             pretrained_cfg_overlay=pretrained_cfg_overlay,
             cache_dir=cache_dir,
             **kwargs,
-        )
+        )  # 存在 *args/**kwargs，未转换，需手动确认参数映射;
 
     if checkpoint_path:
         load_checkpoint(model, checkpoint_path)
