--- pytorch+++ mindspore@@ -1,9 +1,12 @@+import mindspore as ms
+import mindspore.nn as msnn
+import mindspore.ops as msops
+import mindspore.mint as mint
+from mindspore.mint import nn, ops
 import fnmatch
 import logging
 from itertools import islice
 from typing import Collection, Optional
-
-from torch import nn as nn
 
 from timm.models import group_parameters
 
@@ -17,7 +20,7 @@ 
 
 def param_groups_weight_decay(
-        model: nn.Module,
+        model: msnn.Cell,
         weight_decay: float = 1e-5,
         no_weight_decay_list: Collection[str] = (),
         fallback_list: Collection[str] = (),
@@ -100,7 +103,7 @@ 
 
 def param_groups_layer_decay(
-        model: nn.Module,
+        model: msnn.Cell,
         weight_decay: float = 0.05,
         no_weight_decay_list: Collection[str] = (),
         fallback_list: Collection[str] = (),
