--- pytorch+++ mindspore@@ -1,12 +1,15 @@+import mindspore as ms
+import mindspore.nn as msnn
+import mindspore.ops as msops
+import mindspore.mint as mint
+from mindspore.mint import nn, ops
 from typing import Callable, Dict, List, Optional, Union, Tuple, Type
-
-import torch
-from torch import nn
+# from torch import nn
 
 try:
     # NOTE we wrap torchvision fns to use timm leaf / no trace definitions
-    from torchvision.models.feature_extraction import create_feature_extractor as _create_feature_extractor
-    from torchvision.models.feature_extraction import get_graph_node_names as _get_graph_node_names
+    # from torchvision.models.feature_extraction import create_feature_extractor as _create_feature_extractor
+    # from torchvision.models.feature_extraction import get_graph_node_names as _get_graph_node_names
     has_fx_feature_extraction = True
 except ImportError:
     has_fx_feature_extraction = False
@@ -60,6 +63,7 @@     return list(_autowrap_functions)
 
 
+# 类型标注 'torch.nn.Module' 未在映射表(api_mapping_out_excel.json)中找到，需手动确认;
 def get_graph_node_names(model: nn.Module) -> Tuple[List[str], List[str]]:
     return _get_graph_node_names(
         model,
@@ -67,9 +71,10 @@             'leaf_modules': list(_leaf_modules),
             'autowrap_functions': list(_autowrap_functions)
         }
-    )
+    )  # 'torchvision.models.feature_extraction.get_graph_node_names' 未在映射表(api_mapping_out_excel.json)中找到，需手动确认;
 
 
+# 类型标注 'torch.nn.Module' 未在映射表(api_mapping_out_excel.json)中找到，需手动确认;
 def create_feature_extractor(model: nn.Module, return_nodes: Union[Dict[str, str], List[str]]):
     assert has_fx_feature_extraction, 'Please update to PyTorch 1.10+, torchvision 0.11+ for FX feature extraction'
     return _create_feature_extractor(
@@ -78,4 +83,4 @@             'leaf_modules': list(_leaf_modules),
             'autowrap_functions': list(_autowrap_functions)
         }
-    )+    )  # 'torchvision.models.feature_extraction.create_feature_extractor' 未在映射表(api_mapping_out_excel.json)中找到，需手动确认;