--- pytorch+++ mindspore@@ -1,3 +1,8 @@+import mindspore as ms
+import mindspore.nn as msnn
+import mindspore.ops as msops
+import mindspore.mint as mint
+from mindspore.mint import nn, ops
 """ NormAct (Normalization + Activation Layer) Factory
 
 Create norm + act combo modules that attempt to be backwards compatible with separate norm + act
@@ -100,7 +105,7 @@     layer = get_norm_act_layer(layer_name, act_layer=act_layer)
     layer_instance = layer(num_features, apply_act=apply_act, **kwargs)
     if jit:
-        layer_instance = torch.jit.script(layer_instance)
+        layer_instance = torch.jit.script(layer_instance)  # 'torch.jit.script' 未在映射表(api_mapping_out_excel.json)中找到，需手动确认;
     return layer_instance
 
 
