--- pytorch+++ mindspore@@ -1,9 +1,14 @@+import mindspore as ms
+import mindspore.nn as msnn
+import mindspore.ops as msops
+import mindspore.mint as mint
+from mindspore.mint import nn, ops
 """ Interpolation helpers for timm layers
 
 RegularGridInterpolator from https://github.com/sbarratt/torch_interpolations
 Copyright Shane Barratt, Apache 2.0 license
 """
-import torch
+# import torch
 from itertools import product
 
 
@@ -44,7 +49,7 @@         dists = []
         overalls = []
         for p, x in zip(self.points, points_to_interp):
-            idx_right = torch.bucketize(x, p)
+            idx_right = torch.bucketize(x, p)  # 'torch.bucketize' 未在映射表(api_mapping_out_excel.json)中找到，需手动确认;
             idx_right[idx_right >= p.shape[0]] = p.shape[0] - 1
             idx_left = (idx_right - 1).clamp(0, p.shape[0] - 1)
             dist_left = x - p[idx_left]
@@ -63,6 +68,6 @@             as_s = [idx[onoff] for onoff, idx in zip(indexer, idxs)]
             bs_s = [dist[1 - onoff] for onoff, dist in zip(indexer, dists)]
             numerator += self.values[as_s] * \
-                torch.prod(torch.stack(bs_s), dim=0)
-        denominator = torch.prod(torch.stack(overalls), dim=0)
+                mint.prod(mint.stack(bs_s))
+        denominator = mint.prod(mint.stack(overalls))
         return numerator / denominator
