--- pytorch+++ mindspore@@ -1,10 +1,13 @@+import mindspore as ms
+import mindspore.nn as msnn
+import mindspore.ops as msops
+import mindspore.mint as mint
+from mindspore.mint import nn, ops
 """ AvgPool2d w/ Same Padding
 
 Hacked together by / Copyright 2020 Ross Wightman
 """
-import torch
-import torch.nn as nn
-import torch.nn.functional as F
+# import torch.nn as nn
 from typing import List, Tuple, Optional, Union
 
 from ._fx import register_notrace_module
@@ -13,7 +16,7 @@ 
 
 def avg_pool2d_same(
-        x: torch.Tensor,
+        x: ms.Tensor,
         kernel_size: List[int],
         stride: List[int],
         padding: List[int] = (0, 0),
@@ -22,7 +25,7 @@ ):
     # FIXME how to deal with count_include_pad vs not for external padding?
     x = pad_same(x, kernel_size, stride)
-    return F.avg_pool2d(x, kernel_size, stride, (0, 0), ceil_mode, count_include_pad)
+    return nn.functional.avg_pool2d(x, kernel_size, stride, (0, 0), ceil_mode, count_include_pad)
 
 
 @register_notrace_module
@@ -43,12 +46,12 @@ 
     def forward(self, x):
         x = pad_same(x, self.kernel_size, self.stride)
-        return F.avg_pool2d(
+        return nn.functional.avg_pool2d(
             x, self.kernel_size, self.stride, self.padding, self.ceil_mode, self.count_include_pad)
 
 
 def max_pool2d_same(
-        x: torch.Tensor,
+        x: ms.Tensor,
         kernel_size: List[int],
         stride: List[int],
         padding: List[int] = (0, 0),
@@ -56,7 +59,7 @@         ceil_mode: bool = False,
 ):
     x = pad_same(x, kernel_size, stride, value=-float('inf'))
-    return F.max_pool2d(x, kernel_size, stride, (0, 0), dilation, ceil_mode)
+    return nn.functional.max_pool2d(x, kernel_size, stride, (0, 0), dilation, ceil_mode)
 
 
 @register_notrace_module
@@ -78,7 +81,7 @@ 
     def forward(self, x):
         x = pad_same(x, self.kernel_size, self.stride, value=-float('inf'))
-        return F.max_pool2d(x, self.kernel_size, self.stride, (0, 0), self.dilation, self.ceil_mode)
+        return nn.functional.max_pool2d(x, self.kernel_size, self.stride, (0, 0), self.dilation, self.ceil_mode)
 
 
 def create_pool2d(pool_type, kernel_size, stride=None, **kwargs):
