--- pytorch+++ mindspore@@ -1,13 +1,18 @@+import mindspore as ms
+import mindspore.nn as msnn
+import mindspore.ops as msops
+import mindspore.mint as mint
+from mindspore.mint import nn, ops
 from contextlib import nullcontext
 from functools import wraps
 from typing import Callable, Optional, Tuple, Type, TypeVar, Union, overload, ContextManager
 
-import torch
+# import torch
 
 __all__ = ["LayerType", "PadType", "nullwrap", "disable_compiler"]
 
 
-LayerType = Union[str, Callable, Type[torch.nn.Module]]
+LayerType = Union[str, Callable, Type[msnn.Cell]]
 PadType = Union[str, int, Tuple[int, int]]
 
 F = TypeVar("F", bound=Callable[..., object])
@@ -27,7 +32,7 @@     # as a decorator
     @wraps(fn)
     def wrapper(*args, **kwargs):
-        return fn(*args, **kwargs)
+        return fn(*args, **kwargs)  # 存在 *args/**kwargs，未转换，需手动确认参数映射;
     return wrapper  # `@nullwrap`
 
 
