--- pytorch+++ mindspore@@ -1,5 +1,9 @@-import torch
-from torch.fx import symbolic_trace, Tracer, Graph, GraphModule, Node
+import mindspore as ms
+import mindspore.nn as msnn
+import mindspore.ops as msops
+import mindspore.mint as mint
+from mindspore.mint import nn, ops
+# from torch.fx import symbolic_trace, Tracer, Graph, GraphModule, Node
 from typing import Any, Callable, Dict, Optional, Tuple, Union
 
 
@@ -42,15 +46,15 @@ for `ReLU`.
 """
 
-class M1(torch.nn.Module):
+class M1(msnn.Cell):
     def __init__(self):
         super().__init__()
-        self.relu = torch.nn.ReLU()
+        self.relu = nn.ReLU()
 
-    def forward(self, x):
+    def construct(self, x):
         return self.relu(x)
 
-default_traced: GraphModule = symbolic_trace(M1())
+default_traced: GraphModule = symbolic_trace(M1())  # 'torch.fx.GraphModule' 未在映射表(api_mapping_out_excel.json)中找到，需手动确认;; 'torch.fx.symbolic_trace' 未在映射表(api_mapping_out_excel.json)中找到，需手动确认;
 """
 Tracing with the default tracer and calling `print_tabular` produces:
 
@@ -63,9 +67,10 @@ """
 default_traced.graph.print_tabular()
 
+# 'torch.fx.Tracer' 未在映射表(api_mapping_out_excel.json)中找到，需手动确认;
 class LowerReluTracer(Tracer):
-    def is_leaf_module(self, m : torch.nn.Module, qualname : str):
-        if isinstance(m, torch.nn.ReLU):
+    def is_leaf_module(self, m : msnn.Cell, qualname : str):
+        if isinstance(m, nn.ReLU):
             return False
         return super().is_leaf_module(m, qualname)
 
@@ -79,7 +84,7 @@     output         output  output                             (relu_1,)  {}
 """
 lower_relu_tracer = LowerReluTracer()
-custom_traced_graph: Graph = lower_relu_tracer.trace(M1())
+custom_traced_graph: Graph = lower_relu_tracer.trace(M1())  # 'torch.fx.Graph' 未在映射表(api_mapping_out_excel.json)中找到，需手动确认;
 custom_traced_graph.print_tabular()
 
 
@@ -91,11 +96,13 @@ each Node during its creation
 """
 
-class M2(torch.nn.Module):
-    def forward(self, a, b):
+class M2(msnn.Cell):
+    def construct(self, a, b):
         return a + b
 
+# 'torch.fx.Tracer' 未在映射表(api_mapping_out_excel.json)中找到，需手动确认;
 class TaggingTracer(Tracer):
+    # 'torch.fx.Node' 未在映射表(api_mapping_out_excel.json)中找到，需手动确认;
     def create_node(self, kind : str, target : Union[str, Callable],
                     args : Tuple[Any], kwargs : Dict[str, Any], name : Optional[str] = None,
                     type_expr : Optional[Any] = None) -> Node:
@@ -103,8 +110,9 @@         n.tag = "foo"
         return n
 
-custom_traced_graph: Graph = TaggingTracer().trace(M2())
+custom_traced_graph: Graph = TaggingTracer().trace(M2())  # 'torch.fx.Graph' 未在映射表(api_mapping_out_excel.json)中找到，需手动确认;
 
+# 'torch.fx.Graph' 未在映射表(api_mapping_out_excel.json)中找到，需手动确认;
 def assert_all_nodes_have_tags(g: Graph) -> bool:
     for n in g.nodes:
         if not hasattr(n, "tag") or not n.tag == "foo":
