--- pytorch+++ mindspore@@ -1,5 +1,9 @@-import torch
-from torch.fx import symbolic_trace
+import mindspore as ms
+import mindspore.nn as msnn
+import mindspore.ops as msops
+import mindspore.mint as mint
+from mindspore.mint import nn, ops
+# from torch.fx import symbolic_trace
 import operator
 
 """
@@ -29,12 +33,12 @@ """
 
 # Sample module
-class M(torch.nn.Module):
-    def forward(self, x, y):
-        return x + y, torch.add(x, y), x.add(y)
+class M(msnn.Cell):
+    def construct(self, x, y):
+        return x + y, mint.add(x, y), x.add(y)
 
 # Symbolically trace an instance of the module
-traced = symbolic_trace(M())
+traced = symbolic_trace(M())  # 'torch.fx.symbolic_trace' 未在映射表(api_mapping_out_excel.json)中找到，需手动确认;
 
 # As demonstrated in the above example, there are several different ways
 # to denote addition. The possible cases are:
@@ -45,7 +49,7 @@ #     3. `x.add(y)` - The Tensor method call, whose target we can match
 #         as a string.
 
-patterns = set([operator.add, torch.add, "add"])
+patterns = set([operator.add, mint.add, "add"])
 
 # Go through all the nodes in the Graph
 for n in traced.graph.nodes:
@@ -54,7 +58,7 @@         # Set the insert point, add the new node, and replace all uses
         # of `n` with the new node
         with traced.graph.inserting_after(n):
-            new_node = traced.graph.call_function(torch.bitwise_and, n.args, n.kwargs)
+            new_node = traced.graph.call_function(mint.bitwise_and, n.args, n.kwargs)
             n.replace_all_uses_with(new_node)
         # Remove the old node from the graph
         traced.graph.erase_node(n)
