--- pytorch+++ mindspore@@ -1,5 +1,8 @@-import torch
-from torch import nn
+import mindspore as ms
+import mindspore.nn as msnn
+import mindspore.ops as msops
+import mindspore.mint as mint
+from mindspore.mint import nn, ops
 
 def exists(val):
     return val is not None
@@ -15,7 +18,7 @@         return tuple(map(fn, val))
     return fn(val)
 
-class Extractor(nn.Module):
+class Extractor(msnn.Cell):
     def __init__(
         self,
         vit,
@@ -69,7 +72,7 @@         del self.latents
         self.latents = None
 
-    def forward(
+    def construct(
         self,
         img,
         return_embeddings_only = False
