--- pytorch+++ mindspore@@ -1,9 +1,13 @@+import mindspore as ms
+import mindspore.nn as msnn
+import mindspore.ops as msops
+import mindspore.mint as mint
+from mindspore.mint import nn, ops
 from functools import partial
-import torch
-from torch import nn, einsum
+# from torch import nn, einsum
 
 from einops import rearrange
-from einops.layers.torch import Rearrange, Reduce
+# from einops.layers.torch import Rearrange, Reduce
 
 # helpers
 
@@ -12,33 +16,34 @@ 
 # classes
 
-class LayerNorm(nn.Module):
+class LayerNorm(msnn.Cell):
     def __init__(self, dim, eps = 1e-5):
         super().__init__()
         self.eps = eps
-        self.g = nn.Parameter(torch.ones(1, dim, 1, 1))
-        self.b = nn.Parameter(torch.zeros(1, dim, 1, 1))
+        self.g = ms.Parameter(mint.ones(size = (1, dim, 1, 1)))
+        self.b = ms.Parameter(mint.zeros(size = (1, dim, 1, 1)))
 
-    def forward(self, x):
-        var = torch.var(x, dim = 1, unbiased = False, keepdim = True)
-        mean = torch.mean(x, dim = 1, keepdim = True)
+    def construct(self, x):
+        var = mint.var(x, dim = 1, keepdim = True)
+        mean = mint.mean(x, dim = 1, keepdim = True)
         return (x - mean) / (var + self.eps).sqrt() * self.g + self.b
 
-class FeedForward(nn.Module):
+class FeedForward(msnn.Cell):
     def __init__(self, dim, mlp_mult = 4, dropout = 0.):
         super().__init__()
-        self.net = nn.Sequential(
+        self.net = msnn.SequentialCell(
+            [
             LayerNorm(dim),
             nn.Conv2d(dim, dim * mlp_mult, 1),
             nn.GELU(),
             nn.Dropout(dropout),
             nn.Conv2d(dim * mlp_mult, dim, 1),
             nn.Dropout(dropout)
-        )
-    def forward(self, x):
+        ])
+    def construct(self, x):
         return self.net(x)
 
-class Attention(nn.Module):
+class Attention(msnn.Cell):
     def __init__(self, dim, heads = 8, dropout = 0.):
         super().__init__()
         dim_head = dim // heads
@@ -51,12 +56,13 @@         self.dropout = nn.Dropout(dropout)
         self.to_qkv = nn.Conv2d(dim, inner_dim * 3, 1, bias = False)
 
-        self.to_out = nn.Sequential(
+        self.to_out = msnn.SequentialCell(
+            [
             nn.Conv2d(inner_dim, dim, 1),
             nn.Dropout(dropout)
-        )
+        ])
 
-    def forward(self, x):
+    def construct(self, x):
         b, c, h, w, heads = *x.shape, self.heads
 
         x = self.norm(x)
@@ -64,34 +70,35 @@         qkv = self.to_qkv(x).chunk(3, dim = 1)
         q, k, v = map(lambda t: rearrange(t, 'b (h d) x y -> b h (x y) d', h = heads), qkv)
 
-        dots = einsum('b h i d, b h j d -> b h i j', q, k) * self.scale
+        dots = mint.einsum('b h i d, b h j d -> b h i j', q, k) * self.scale
 
         attn = self.attend(dots)
         attn = self.dropout(attn)
 
-        out = einsum('b h i j, b h j d -> b h i d', attn, v)
+        out = mint.einsum('b h i j, b h j d -> b h i d', attn, v)
         out = rearrange(out, 'b h (x y) d -> b (h d) x y', x = h, y = w)
         return self.to_out(out)
 
 def Aggregate(dim, dim_out):
-    return nn.Sequential(
+    return msnn.SequentialCell(
+        [
         nn.Conv2d(dim, dim_out, 3, padding = 1),
         LayerNorm(dim_out),
         nn.MaxPool2d(3, stride = 2, padding = 1)
-    )
+    ])
 
-class Transformer(nn.Module):
+class Transformer(msnn.Cell):
     def __init__(self, dim, seq_len, depth, heads, mlp_mult, dropout = 0.):
         super().__init__()
-        self.layers = nn.ModuleList([])
-        self.pos_emb = nn.Parameter(torch.randn(seq_len))
+        self.layers = msnn.CellList([])
+        self.pos_emb = ms.Parameter(mint.randn(seq_len))
 
         for _ in range(depth):
-            self.layers.append(nn.ModuleList([
+            self.layers.append(msnn.CellList([
                 Attention(dim, heads = heads, dropout = dropout),
                 FeedForward(dim, mlp_mult, dropout = dropout)
             ]))
-    def forward(self, x):
+    def construct(self, x):
         *_, h, w = x.shape
 
         pos_emb = self.pos_emb[:(h * w)]
@@ -103,7 +110,7 @@             x = ff(x) + x
         return x
 
-class NesT(nn.Module):
+class NesT(msnn.Cell):
     def __init__(
         self,
         *,
@@ -137,34 +144,36 @@         layer_dims = [*layer_dims, layer_dims[-1]]
         dim_pairs = zip(layer_dims[:-1], layer_dims[1:])
 
-        self.to_patch_embedding = nn.Sequential(
+        self.to_patch_embedding = msnn.SequentialCell(
+            [
             Rearrange('b c (h p1) (w p2) -> b (p1 p2 c) h w', p1 = patch_size, p2 = patch_size),
             LayerNorm(patch_dim),
             nn.Conv2d(patch_dim, layer_dims[0], 1),
             LayerNorm(layer_dims[0])
-        )
+        ])
 
         block_repeats = cast_tuple(block_repeats, num_hierarchies)
 
-        self.layers = nn.ModuleList([])
+        self.layers = msnn.CellList([])
 
         for level, heads, (dim_in, dim_out), block_repeat in zip(hierarchies, layer_heads, dim_pairs, block_repeats):
             is_last = level == 0
             depth = block_repeat
 
-            self.layers.append(nn.ModuleList([
+            self.layers.append(msnn.CellList([
                 Transformer(dim_in, seq_len, depth, heads, mlp_mult, dropout),
-                Aggregate(dim_in, dim_out) if not is_last else nn.Identity()
+                Aggregate(dim_in, dim_out) if not is_last else msnn.Identity()
             ]))
 
 
-        self.mlp_head = nn.Sequential(
+        self.mlp_head = msnn.SequentialCell(
+            [
             LayerNorm(last_dim),
             Reduce('b c h w -> b c', 'mean'),
             nn.Linear(last_dim, num_classes)
-        )
+        ])
 
-    def forward(self, img):
+    def construct(self, img):
         x = self.to_patch_embedding(img)
         b, c, h, w = x.shape
 
