--- pytorch+++ mindspore@@ -1,11 +1,13 @@+import mindspore as ms
+import mindspore.nn as msnn
+import mindspore.ops as msops
+import mindspore.mint as mint
+from mindspore.mint import nn, ops
 from math import sqrt
-
-import torch
-from torch import nn, einsum
-import torch.nn.functional as F
+# from torch import nn, einsum
 
 from einops import rearrange, repeat
-from einops.layers.torch import Rearrange
+# from einops.layers.torch import Rearrange
 
 # helpers
 
@@ -17,21 +19,22 @@ 
 # classes
 
-class FeedForward(nn.Module):
+class FeedForward(msnn.Cell):
     def __init__(self, dim, hidden_dim, dropout = 0.):
         super().__init__()
-        self.net = nn.Sequential(
+        self.net = msnn.SequentialCell(
+            [
             nn.LayerNorm(dim),
             nn.Linear(dim, hidden_dim),
             nn.GELU(),
             nn.Dropout(dropout),
             nn.Linear(hidden_dim, dim),
             nn.Dropout(dropout)
-        )
-    def forward(self, x):
+        ])
+    def construct(self, x):
         return self.net(x)
 
-class Attention(nn.Module):
+class Attention(msnn.Cell):
     def __init__(self, dim, heads = 8, dim_head = 64, dropout = 0.):
         super().__init__()
         inner_dim = dim_head *  heads
@@ -45,37 +48,38 @@         self.dropout = nn.Dropout(dropout)
         self.to_qkv = nn.Linear(dim, inner_dim * 3, bias = False)
 
-        self.to_out = nn.Sequential(
+        self.to_out = msnn.SequentialCell(
+            [
             nn.Linear(inner_dim, dim),
             nn.Dropout(dropout)
-        ) if project_out else nn.Identity()
+        ]) if project_out else msnn.Identity()
 
-    def forward(self, x):
+    def construct(self, x):
         b, n, _, h = *x.shape, self.heads
 
         x = self.norm(x)
         qkv = self.to_qkv(x).chunk(3, dim = -1)
         q, k, v = map(lambda t: rearrange(t, 'b n (h d) -> b h n d', h = h), qkv)
 
-        dots = einsum('b h i d, b h j d -> b h i j', q, k) * self.scale
+        dots = mint.einsum('b h i d, b h j d -> b h i j', q, k) * self.scale
 
         attn = self.attend(dots)
         attn = self.dropout(attn)
 
-        out = einsum('b h i j, b h j d -> b h i d', attn, v)
+        out = mint.einsum('b h i j, b h j d -> b h i d', attn, v)
         out = rearrange(out, 'b h n d -> b n (h d)')
         return self.to_out(out)
 
-class Transformer(nn.Module):
+class Transformer(msnn.Cell):
     def __init__(self, dim, depth, heads, dim_head, mlp_dim, dropout = 0.):
         super().__init__()
-        self.layers = nn.ModuleList([])
+        self.layers = msnn.CellList([])
         for _ in range(depth):
-            self.layers.append(nn.ModuleList([
+            self.layers.append(msnn.CellList([
                 Attention(dim, heads = heads, dim_head = dim_head, dropout = dropout),
                 FeedForward(dim, mlp_dim, dropout = dropout)
             ]))
-    def forward(self, x):
+    def construct(self, x):
         for attn, ff in self.layers:
             x = attn(x) + x
             x = ff(x) + x
@@ -83,25 +87,26 @@ 
 # depthwise convolution, for pooling
 
-class DepthWiseConv2d(nn.Module):
+class DepthWiseConv2d(msnn.Cell):
     def __init__(self, dim_in, dim_out, kernel_size, padding, stride, bias = True):
         super().__init__()
-        self.net = nn.Sequential(
-            nn.Conv2d(dim_in, dim_out, kernel_size = kernel_size, padding = padding, groups = dim_in, stride = stride, bias = bias),
+        self.net = msnn.SequentialCell(
+            [
+            nn.Conv2d(dim_in, dim_out, kernel_size = kernel_size, stride = stride, padding = padding, groups = dim_in, bias = bias),
             nn.Conv2d(dim_out, dim_out, kernel_size = 1, bias = bias)
-        )
-    def forward(self, x):
+        ])
+    def construct(self, x):
         return self.net(x)
 
 # pooling layer
 
-class Pool(nn.Module):
+class Pool(msnn.Cell):
     def __init__(self, dim):
         super().__init__()
         self.downsample = DepthWiseConv2d(dim, dim * 2, kernel_size = 3, stride = 2, padding = 1)
         self.cls_ff = nn.Linear(dim, dim * 2)
 
-    def forward(self, x):
+    def construct(self, x):
         cls_token, tokens = x[:, :1], x[:, 1:]
 
         cls_token = self.cls_ff(cls_token)
@@ -110,11 +115,11 @@         tokens = self.downsample(tokens)
         tokens = rearrange(tokens, 'b c h w -> b (h w) c')
 
-        return torch.cat((cls_token, tokens), dim = 1)
+        return mint.cat((cls_token, tokens), dim = 1)
 
 # main class
 
-class PiT(nn.Module):
+class PiT(msnn.Cell):
     def __init__(
         self,
         *,
@@ -137,17 +142,18 @@ 
         patch_dim = channels * patch_size ** 2
 
-        self.to_patch_embedding = nn.Sequential(
+        self.to_patch_embedding = msnn.SequentialCell(
+            [
             nn.Unfold(kernel_size = patch_size, stride = patch_size // 2),
             Rearrange('b c n -> b n c'),
             nn.Linear(patch_dim, dim)
-        )
+        ])
 
         output_size = conv_output_size(image_size, patch_size, patch_size // 2)
         num_patches = output_size ** 2
 
-        self.pos_embedding = nn.Parameter(torch.randn(1, num_patches + 1, dim))
-        self.cls_token = nn.Parameter(torch.randn(1, 1, dim))
+        self.pos_embedding = ms.Parameter(mint.randn(size = (1, num_patches + 1, dim)))
+        self.cls_token = ms.Parameter(mint.randn(size = (1, 1, dim)))
         self.dropout = nn.Dropout(emb_dropout)
 
         layers = []
@@ -161,19 +167,22 @@                 layers.append(Pool(dim))
                 dim *= 2
 
-        self.layers = nn.Sequential(*layers)
+        self.layers = msnn.SequentialCell([
+            layers
+        ])
 
-        self.mlp_head = nn.Sequential(
+        self.mlp_head = msnn.SequentialCell(
+            [
             nn.LayerNorm(dim),
             nn.Linear(dim, num_classes)
-        )
+        ])
 
-    def forward(self, img):
+    def construct(self, img):
         x = self.to_patch_embedding(img)
         b, n, _ = x.shape
 
         cls_tokens = repeat(self.cls_token, '() n d -> b n d', b = b)
-        x = torch.cat((cls_tokens, x), dim=1)
+        x = mint.cat((cls_tokens, x), dim = 1)
         x += self.pos_embedding[:, :n+1]
         x = self.dropout(x)
 
