--- pytorch+++ mindspore@@ -1,13 +1,17 @@+import mindspore as ms
+import mindspore.nn as msnn
+import mindspore.ops as msops
+import mindspore.mint as mint
+from mindspore.mint import nn, ops
 from packaging import version
 from collections import namedtuple
 
-import torch
-from torch import nn
-import torch.nn.functional as F
-from torch.nn import Module, ModuleList
+# import torch
+# from torch import nn
+# import torch.nn.functional as F
 
 from einops import rearrange
-from einops.layers.torch import Rearrange
+# from einops.layers.torch import Rearrange
 
 # constants
 
@@ -18,32 +22,29 @@ def pair(t):
     return t if isinstance(t, tuple) else (t, t)
 
-def posemb_sincos_3d(patches, temperature = 10000, dtype = torch.float32):
+def posemb_sincos_3d(patches, temperature = 10000, dtype = ms.float32):
     _, f, h, w, dim, device, dtype = *patches.shape, patches.device, patches.dtype
 
-    z, y, x = torch.meshgrid(
-        torch.arange(f, device = device),
-        torch.arange(h, device = device),
-        torch.arange(w, device = device),
-    indexing = 'ij')
+    z, y, x = mint.meshgrid(
+        mint.arange(f), mint.arange(h), mint.arange(w), indexing = 'ij')  # 'torch.arange':没有对应的mindspore参数 'device' (position 6);
 
     fourier_dim = dim // 6
 
-    omega = torch.arange(fourier_dim, device = device) / (fourier_dim - 1)
+    omega = mint.arange(fourier_dim) / (fourier_dim - 1)  # 'torch.arange':没有对应的mindspore参数 'device' (position 6);
     omega = 1. / (temperature ** omega)
 
     z = z.flatten()[:, None] * omega[None, :]
     y = y.flatten()[:, None] * omega[None, :]
     x = x.flatten()[:, None] * omega[None, :] 
 
-    pe = torch.cat((x.sin(), x.cos(), y.sin(), y.cos(), z.sin(), z.cos()), dim = 1)
+    pe = mint.cat((x.sin(), x.cos(), y.sin(), y.cos(), z.sin(), z.cos()), dim = 1)
 
-    pe = F.pad(pe, (0, dim - (fourier_dim * 6))) # pad if feature dimension not cleanly divisible by 6
+    pe = nn.functional.pad(pe, (0, dim - (fourier_dim * 6))) # pad if feature dimension not cleanly divisible by 6
     return pe.type(dtype)
 
 # main class
 
-class Attend(Module):
+class Attend(msnn.Cell):
     def __init__(self, use_flash = False, config: Config = Config(True, True, True)):
         super().__init__()
         self.config = config
@@ -58,7 +59,7 @@ 
         return out
 
-    def forward(self, q, k, v):
+    def construct(self, q, k, v):
         n, device, scale = q.shape[-2], q.device, q.shape[-1] ** -0.5
 
         if self.use_flash:
@@ -80,19 +81,20 @@ 
 # classes
 
-class FeedForward(Module):
+class FeedForward(msnn.Cell):
     def __init__(self, dim, hidden_dim):
         super().__init__()
-        self.net = nn.Sequential(
+        self.net = msnn.SequentialCell(
+            [
             nn.LayerNorm(dim),
             nn.Linear(dim, hidden_dim),
             nn.GELU(),
-            nn.Linear(hidden_dim, dim),
-        )
-    def forward(self, x):
+            nn.Linear(hidden_dim, dim)
+        ])
+    def construct(self, x):
         return self.net(x)
 
-class Attention(Module):
+class Attention(msnn.Cell):
     def __init__(self, dim, heads = 8, dim_head = 64, use_flash = True):
         super().__init__()
         inner_dim = dim_head *  heads
@@ -105,7 +107,7 @@         self.to_qkv = nn.Linear(dim, inner_dim * 3, bias = False)
         self.to_out = nn.Linear(inner_dim, dim, bias = False)
 
-    def forward(self, x):
+    def construct(self, x):
         x = self.norm(x)
 
         qkv = self.to_qkv(x).chunk(3, dim = -1)
@@ -116,24 +118,24 @@         out = rearrange(out, 'b h n d -> b n (h d)')
         return self.to_out(out)
 
-class Transformer(Module):
+class Transformer(msnn.Cell):
     def __init__(self, dim, depth, heads, dim_head, mlp_dim, use_flash):
         super().__init__()
-        self.layers = ModuleList([])
+        self.layers = msnn.CellList([])
         for _ in range(depth):
-            self.layers.append(ModuleList([
+            self.layers.append(msnn.CellList([
                 Attention(dim, heads = heads, dim_head = dim_head, use_flash = use_flash),
                 FeedForward(dim, mlp_dim)
             ]))
 
-    def forward(self, x):
+    def construct(self, x):
         for attn, ff in self.layers:
             x = attn(x) + x
             x = ff(x) + x
 
         return x
 
-class SimpleViT(Module):
+class SimpleViT(msnn.Cell):
     def __init__(self, *, image_size, image_patch_size, frames, frame_patch_size, num_classes, dim, depth, heads, mlp_dim, channels = 3, dim_head = 64, use_flash_attn = True):
         super().__init__()
         image_height, image_width = pair(image_size)
@@ -145,19 +147,20 @@         num_patches = (image_height // patch_height) * (image_width // patch_width) * (frames // frame_patch_size)
         patch_dim = channels * patch_height * patch_width * frame_patch_size
 
-        self.to_patch_embedding = nn.Sequential(
+        self.to_patch_embedding = msnn.SequentialCell(
+            [
             Rearrange('b c (f pf) (h p1) (w p2) -> b f h w (pf p1 p2 c)', p1 = patch_height, p2 = patch_width, pf = frame_patch_size),
             nn.LayerNorm(patch_dim),
             nn.Linear(patch_dim, dim),
-            nn.LayerNorm(dim),
-        )
+            nn.LayerNorm(dim)
+        ])
 
         self.transformer = Transformer(dim, depth, heads, dim_head, mlp_dim, use_flash_attn)
 
-        self.to_latent = nn.Identity()
+        self.to_latent = msnn.Identity()
         self.linear_head = nn.Linear(dim, num_classes)
 
-    def forward(self, video):
+    def construct(self, video):
         *_, h, w, dtype = *video.shape, video.dtype
 
         x = self.to_patch_embedding(video)
