--- pytorch+++ mindspore@@ -1,9 +1,12 @@-import torch
-from torch import nn
-from torch.nn import Module, ModuleList
+import mindspore as ms
+import mindspore.nn as msnn
+import mindspore.ops as msops
+import mindspore.mint as mint
+from mindspore.mint import nn, ops
+# from torch import nn
 
 from einops import rearrange
-from einops.layers.torch import Rearrange
+# from einops.layers.torch import Rearrange
 
 # helpers
 
@@ -16,28 +19,29 @@ def pair(t):
     return t if isinstance(t, tuple) else (t, t)
 
-def posemb_sincos_2d(h, w, dim, temperature: int = 10000, dtype = torch.float32):
-    y, x = torch.meshgrid(torch.arange(h), torch.arange(w), indexing="ij")
+def posemb_sincos_2d(h, w, dim, temperature: int = 10000, dtype = ms.float32):
+    y, x = mint.meshgrid(mint.arange(h), mint.arange(w), indexing = "ij")
     assert (dim % 4) == 0, "feature dimension must be multiple of 4 for sincos emb"
-    omega = torch.arange(dim // 4) / (dim // 4 - 1)
+    omega = mint.arange(dim // 4) / (dim // 4 - 1)
     omega = 1.0 / (temperature ** omega)
 
     y = y.flatten()[:, None] * omega[None, :]
     x = x.flatten()[:, None] * omega[None, :]
-    pe = torch.cat((x.sin(), x.cos(), y.sin(), y.cos()), dim=1)
+    pe = mint.cat((x.sin(), x.cos(), y.sin(), y.cos()), dim = 1)
     return pe.type(dtype)
 
 # classes
 
 def FeedForward(dim, hidden_dim):
-    return nn.Sequential(
+    return msnn.SequentialCell(
+        [
         nn.LayerNorm(dim),
         nn.Linear(dim, hidden_dim),
         nn.GELU(),
-        nn.Linear(hidden_dim, dim),
-    )
+        nn.Linear(hidden_dim, dim)
+    ])
 
-class Attention(Module):
+class Attention(msnn.Cell):
     def __init__(self, dim, heads = 8, dim_head = 64, learned_value_residual_mix = False):
         super().__init__()
         inner_dim = dim_head *  heads
@@ -50,13 +54,14 @@         self.to_qkv = nn.Linear(dim, inner_dim * 3, bias = False)
         self.to_out = nn.Linear(inner_dim, dim, bias = False)
 
-        self.to_residual_mix = nn.Sequential(
+        self.to_residual_mix = msnn.SequentialCell(
+            [
             nn.Linear(dim, heads),
             nn.Sigmoid(),
             Rearrange('b n h -> b h n 1')
-        ) if learned_value_residual_mix else (lambda _: 0.5)
+        ]) if learned_value_residual_mix else (lambda _: 0.5)
 
-    def forward(self, x, value_residual = None):
+    def construct(self, x, value_residual = None):
         x = self.norm(x)
 
         qkv = self.to_qkv(x).chunk(3, dim = -1)
@@ -66,27 +71,27 @@             mix = self.to_residual_mix(x)
             v = v * mix + value_residual * (1. - mix)
 
-        dots = torch.matmul(q, k.transpose(-1, -2)) * self.scale
+        dots = mint.matmul(q, k.transpose(-1, -2)) * self.scale
 
         attn = self.attend(dots)
 
-        out = torch.matmul(attn, v)
+        out = mint.matmul(attn, v)
         out = rearrange(out, 'b h n d -> b n (h d)')
 
         return self.to_out(out), v
 
-class Transformer(Module):
+class Transformer(msnn.Cell):
     def __init__(self, dim, depth, heads, dim_head, mlp_dim):
         super().__init__()
         self.norm = nn.LayerNorm(dim)
-        self.layers = ModuleList([])
+        self.layers = msnn.CellList([])
         for i in range(depth):
             is_first = i == 0
-            self.layers.append(ModuleList([
+            self.layers.append(msnn.CellList([
                 Attention(dim, heads = heads, dim_head = dim_head, learned_value_residual_mix = not is_first),
                 FeedForward(dim, mlp_dim)
             ]))
-    def forward(self, x):
+    def construct(self, x):
         value_residual = None
 
         for attn, ff in self.layers:
@@ -99,7 +104,7 @@ 
         return self.norm(x)
 
-class SimpleViT(Module):
+class SimpleViT(msnn.Cell):
     def __init__(self, *, image_size, patch_size, num_classes, dim, depth, heads, mlp_dim, channels = 3, dim_head = 64):
         super().__init__()
         image_height, image_width = pair(image_size)
@@ -109,12 +114,13 @@ 
         patch_dim = channels * patch_height * patch_width
 
-        self.to_patch_embedding = nn.Sequential(
+        self.to_patch_embedding = msnn.SequentialCell(
+            [
             Rearrange("b c (h p1) (w p2) -> b (h w) (p1 p2 c)", p1 = patch_height, p2 = patch_width),
             nn.LayerNorm(patch_dim),
             nn.Linear(patch_dim, dim),
-            nn.LayerNorm(dim),
-        )
+            nn.LayerNorm(dim)
+        ])
 
         self.pos_embedding = posemb_sincos_2d(
             h = image_height // patch_height,
@@ -125,11 +131,11 @@         self.transformer = Transformer(dim, depth, heads, dim_head, mlp_dim)
 
         self.pool = "mean"
-        self.to_latent = nn.Identity()
+        self.to_latent = msnn.Identity()
 
         self.linear_head = nn.Linear(dim, num_classes)
 
-    def forward(self, img):
+    def construct(self, img):
         device = img.device
 
         x = self.to_patch_embedding(img)
@@ -154,6 +160,6 @@         mlp_dim = 2048,
     )
 
-    images = torch.randn(2, 3, 256, 256)
+    images = mint.randn(size = (2, 3, 256, 256))
 
     logits = v(images)
