--- pytorch+++ mindspore@@ -1,11 +1,15 @@+import mindspore as ms
+import mindspore.nn as msnn
+import mindspore.ops as msops
+import mindspore.mint as mint
+from mindspore.mint import nn, ops
 import math
-import torch
-from torch import nn
+# from torch import nn
 
-from vit_pytorch.vit import Transformer
+# from vit_pytorch.vit import Transformer
 
 from einops import rearrange, repeat
-from einops.layers.torch import Rearrange
+# from einops.layers.torch import Rearrange
 
 # helpers
 
@@ -17,13 +21,13 @@ 
 # classes
 
-class RearrangeImage(nn.Module):
-    def forward(self, x):
+class RearrangeImage(msnn.Cell):
+    def construct(self, x):
         return rearrange(x, 'b (h w) c -> b c h w', h = int(math.sqrt(x.shape[1])))
 
 # main class
 
-class T2TViT(nn.Module):
+class T2TViT(msnn.Cell):
     def __init__(self, *, image_size, num_classes, dim, depth = None, heads = None, mlp_dim = None, pool = 'cls', channels = 3, dim_head = 64, dropout = 0., emb_dropout = 0., transformer = None, t2t_layers = ((7, 4), (3, 2), (3, 2))):
         super().__init__()
         assert pool in {'cls', 'mean'}, 'pool type must be either cls (cls token) or mean (mean pooling)'
@@ -39,17 +43,17 @@             output_image_size = conv_output_size(output_image_size, kernel_size, stride, stride // 2)
 
             layers.extend([
-                RearrangeImage() if not is_first else nn.Identity(),
-                nn.Unfold(kernel_size = kernel_size, stride = stride, padding = stride // 2),
+                RearrangeImage() if not is_first else msnn.Identity(),
+                nn.Unfold(kernel_size = kernel_size, padding = stride // 2, stride = stride),
                 Rearrange('b c n -> b n c'),
-                Transformer(dim = layer_dim, heads = 1, depth = 1, dim_head = layer_dim, mlp_dim = layer_dim, dropout = dropout) if not is_last else nn.Identity(),
+                Transformer(dim = layer_dim, heads = 1, depth = 1, dim_head = layer_dim, mlp_dim = layer_dim, dropout = dropout) if not is_last else msnn.Identity(),
             ])
 
         layers.append(nn.Linear(layer_dim, dim))
-        self.to_patch_embedding = nn.Sequential(*layers)
+        self.to_patch_embedding = msnn.SequentialCell([layers])
 
-        self.pos_embedding = nn.Parameter(torch.randn(1, output_image_size ** 2 + 1, dim))
-        self.cls_token = nn.Parameter(torch.randn(1, 1, dim))
+        self.pos_embedding = ms.Parameter(mint.randn(size = (1, output_image_size ** 2 + 1, dim)))
+        self.cls_token = ms.Parameter(mint.randn(size = (1, 1, dim)))
         self.dropout = nn.Dropout(emb_dropout)
 
         if not exists(transformer):
@@ -59,16 +63,16 @@             self.transformer = transformer
 
         self.pool = pool
-        self.to_latent = nn.Identity()
+        self.to_latent = msnn.Identity()
 
         self.mlp_head = nn.Linear(dim, num_classes)
 
-    def forward(self, img):
+    def construct(self, img):
         x = self.to_patch_embedding(img)
         b, n, _ = x.shape
 
         cls_tokens = repeat(self.cls_token, '() n d -> b n d', b = b)
-        x = torch.cat((cls_tokens, x), dim=1)
+        x = mint.cat((cls_tokens, x), dim = 1)
         x += self.pos_embedding[:, :n+1]
         x = self.dropout(x)
 
