--- pytorch+++ mindspore@@ -1,10 +1,14 @@+import mindspore as ms
+import mindspore.nn as msnn
+import mindspore.ops as msops
+import mindspore.mint as mint
+from mindspore.mint import nn, ops
 """ PyTorch FX Based Feature Extraction Helpers
 Using https://pytorch.org/vision/stable/feature_extraction.html
 """
 from typing import Callable, Dict, List, Optional, Union, Tuple, Type
 
-import torch
-from torch import nn
+# import torch
 
 from timm.layers import (
     create_feature_extractor,
@@ -35,14 +39,14 @@ ]
 
 
-class FeatureGraphNet(nn.Module):
+class FeatureGraphNet(msnn.Cell):
     """ A FX Graph based feature extractor that works with the model feature_info metadata
     """
     return_dict: torch.jit.Final[bool]
 
     def __init__(
             self,
-            model: nn.Module,
+            model: msnn.Cell,
             out_indices: Tuple[int, ...],
             out_map: Optional[Dict] = None,
             output_fmt: str = 'NCHW',
@@ -57,14 +61,14 @@         self.graph_module = create_feature_extractor(model, return_nodes)
         self.return_dict = return_dict
 
-    def forward(self, x):
+    def construct(self, x):
         out = self.graph_module(x)
         if self.return_dict:
             return out
         return list(out.values())
 
 
-class GraphExtractNet(nn.Module):
+class GraphExtractNet(msnn.Cell):
     """ A standalone feature extraction wrapper that maps dict -> list or single tensor
     NOTE:
       * one can use feature_extractor directly if dictionary output is desired
@@ -82,7 +86,7 @@ 
     def __init__(
             self,
-            model: nn.Module,
+            model: msnn.Cell,
             return_nodes: Union[Dict[str, str], List[str]],
             squeeze_out: bool = True,
             return_dict: bool = False,
@@ -92,7 +96,7 @@         self.graph_module = create_feature_extractor(model, return_nodes)
         self.return_dict = return_dict
 
-    def forward(self, x) -> Union[List[torch.Tensor], torch.Tensor]:
+    def construct(self, x) -> Union[List[ms.Tensor], ms.Tensor]:
         out = self.graph_module(x)
         if self.return_dict:
             return out
