--- pytorch+++ mindspore@@ -1,8 +1,13 @@+import mindspore as ms
+import mindspore.nn as msnn
+import mindspore.ops as msops
+import mindspore.mint as mint
+from mindspore.mint import nn, ops
 import os
 import pkgutil
 from copy import deepcopy
 
-from torch import nn as nn
+# from torch import nn as nn
 
 from timm.layers import Conv2dSame, BatchNormAct2d, Linear
 
@@ -82,7 +87,7 @@             if old_module.groups > 1:
                 in_channels = out_channels
                 g = in_channels
-            new_conv = conv(
+            new_conv = nn.Conv2d(
                 in_channels=in_channels,
                 out_channels=out_channels,
                 kernel_size=old_module.kernel_size,
