--- pytorch+++ mindspore@@ -1,8 +1,13 @@+import mindspore as ms
+import mindspore.nn as msnn
+import mindspore.ops as msops
+import mindspore.mint as mint
+from mindspore.mint import nn, ops
 """ MultiStep LR Scheduler
 
 Basic multi step LR schedule with warmup, noise.
 """
-import torch
+# import torch
 import bisect
 from timm.scheduler.scheduler import Scheduler
 from typing import List
@@ -11,6 +16,7 @@     """
     """
 
+    # 'torch.optim.Optimizer' 未在映射表(api_mapping_out_excel.json)中找到，需手动确认;
     def __init__(
             self,
             optimizer: torch.optim.Optimizer,
