--- pytorch+++ mindspore@@ -1,10 +1,15 @@+import mindspore as ms
+import mindspore.nn as msnn
+import mindspore.ops as msops
+import mindspore.mint as mint
+from mindspore.mint import nn, ops
 """ Plateau Scheduler
 
 Adapts PyTorch plateau scheduler and allows application of noise, warmup.
 
 Hacked together by / Copyright 2020 Ross Wightman
 """
-import torch
+# import torch
 from typing import List, Optional
 
 from .scheduler import Scheduler
@@ -52,7 +57,7 @@             cooldown=cooldown_t,
             mode=mode,
             min_lr=lr_min
-        )
+        )  # 'torch.optim.lr_scheduler.ReduceLROnPlateau' 未在映射表(api_mapping_out_excel.json)中找到，需手动确认;
 
         self.warmup_t = warmup_t
         self.warmup_lr_init = warmup_lr_init
