--- pytorch+++ mindspore@@ -1,3 +1,8 @@+import mindspore as ms
+import mindspore.nn as msnn
+import mindspore.ops as msops
+import mindspore.mint as mint
+from mindspore.mint import nn, ops
 from typing import Any, Dict, Iterable, Union, Protocol, Type
 try:
     from typing import TypeAlias
@@ -8,11 +13,10 @@ except ImportError:
     from typing_extensions import TypeVar
 
-import torch
-import torch.optim
+# import torch
 
 try:
-    from torch.optim.optimizer import ParamsT
+    # from torch.optim.optimizer import ParamsT
 except (ImportError, TypeError):
     ParamsT: TypeAlias = Union[Iterable[torch.Tensor], Iterable[Dict[str, Any]]]
 
@@ -23,6 +27,10 @@ class OptimizerCallable(Protocol):
     """Protocol for optimizer constructor signatures."""
 
+    # 'torch.optim.optimizer.ParamsT' 未在映射表(api_mapping_out_excel.json)中找到，需手动确认;
+    # 'torch.optim.Optimizer' 未在映射表(api_mapping_out_excel.json)中找到，需手动确认;
+    # 'torch.optim' 未在映射表(api_mapping_out_excel.json)中找到，需手动确认;
+    # 'torch' 未在映射表(api_mapping_out_excel.json)中找到，需手动确认;
     def __call__(self, params: ParamsT, **kwargs) -> torch.optim.Optimizer: ...
 
 
