--- pytorch+++ mindspore@@ -1,3 +1,8 @@+import mindspore as ms
+import mindspore.nn as msnn
+import mindspore.ops as msops
+import mindspore.mint as mint
+from mindspore.mint import nn, ops
 """ PyTorch implementation of DualPathNetworks
 Based on original MXNet implementation https://github.com/cypw/DPNs with
 many ideas from another PyTorch implementation https://github.com/oyam/pytorch-DPNs.
@@ -10,9 +15,8 @@ from functools import partial
 from typing import Tuple, Type, Optional
 
-import torch
-import torch.nn as nn
-import torch.nn.functional as F
+# import torch
+# import torch.nn as nn
 
 from timm.data import IMAGENET_DPN_MEAN, IMAGENET_DPN_STD, IMAGENET_DEFAULT_MEAN, IMAGENET_DEFAULT_STD
 from timm.layers import BatchNormAct2d, ConvNormAct, create_conv2d, create_classifier, get_norm_act_layer
@@ -22,35 +26,37 @@ __all__ = ['DPN']
 
 
-class CatBnAct(nn.Module):
+class CatBnAct(msnn.Cell):
     def __init__(
             self,
             in_chs: int,
-            norm_layer: Type[nn.Module] = BatchNormAct2d,
+            norm_layer: Type[msnn.Cell] = BatchNormAct2d,
             device=None,
             dtype=None,
     ):
         dd = {'device': device, 'dtype': dtype}
         super().__init__()
-        self.bn = norm_layer(in_chs, eps=0.001, **dd)
-
+        self.bn = norm_layer(in_chs, eps=0.001, **dd)  # 存在 *args/**kwargs，未转换，需手动确认参数映射;
+
+    # 装饰器 'torch.jit._overload_method' 未在映射表(api_mapping_out_excel.json)中找到，需手动确认;
     @torch.jit._overload_method  # noqa: F811
-    def forward(self, x):
+    def construct(self, x):
         # type: (Tuple[torch.Tensor, torch.Tensor]) -> (torch.Tensor)
         pass
 
+    # 装饰器 'torch.jit._overload_method' 未在映射表(api_mapping_out_excel.json)中找到，需手动确认;
     @torch.jit._overload_method  # noqa: F811
-    def forward(self, x):
+    def construct(self, x):
         # type: (torch.Tensor) -> (torch.Tensor)
         pass
 
-    def forward(self, x):
+    def construct(self, x):
         if isinstance(x, tuple):
-            x = torch.cat(x, dim=1)
+            x = mint.cat(x, dim=1)
         return self.bn(x)
 
 
-class BnActConv2d(nn.Module):
+class BnActConv2d(msnn.Cell):
     def __init__(
             self,
             in_chs: int,
@@ -58,20 +64,20 @@             kernel_size: int,
             stride: int,
             groups: int = 1,
-            norm_layer: Type[nn.Module] = BatchNormAct2d,
+            norm_layer: Type[msnn.Cell] = BatchNormAct2d,
             device=None,
             dtype=None,
     ):
         dd = {'device': device, 'dtype': dtype}
         super().__init__()
-        self.bn = norm_layer(in_chs, eps=0.001, **dd)
-        self.conv = create_conv2d(in_chs, out_chs, kernel_size, stride=stride, groups=groups, **dd)
-
-    def forward(self, x):
+        self.bn = norm_layer(in_chs, eps=0.001, **dd)  # 存在 *args/**kwargs，未转换，需手动确认参数映射;
+        self.conv = create_conv2d(in_chs, out_chs, kernel_size, stride=stride, groups=groups, **dd)  # 存在 *args/**kwargs，未转换，需手动确认参数映射;
+
+    def construct(self, x):
         return self.conv(self.bn(x))
 
 
-class DualPathBlock(nn.Module):
+class DualPathBlock(msnn.Cell):
     def __init__(
             self,
             in_chs: int,
@@ -107,36 +113,38 @@             # Using different member names here to allow easier parameter key matching for conversion
             if self.key_stride == 2:
                 self.c1x1_w_s2 = BnActConv2d(
-                    in_chs=in_chs, out_chs=num_1x1_c + 2 * inc, kernel_size=1, stride=2, **dd)
+                    in_chs=in_chs, out_chs=num_1x1_c + 2 * inc, kernel_size=1, stride=2, **dd)  # 存在 *args/**kwargs，未转换，需手动确认参数映射;
             else:
                 self.c1x1_w_s1 = BnActConv2d(
-                    in_chs=in_chs, out_chs=num_1x1_c + 2 * inc, kernel_size=1, stride=1, **dd)
-
-        self.c1x1_a = BnActConv2d(in_chs=in_chs, out_chs=num_1x1_a, kernel_size=1, stride=1, **dd)
+                    in_chs=in_chs, out_chs=num_1x1_c + 2 * inc, kernel_size=1, stride=1, **dd)  # 存在 *args/**kwargs，未转换，需手动确认参数映射;
+
+        self.c1x1_a = BnActConv2d(in_chs=in_chs, out_chs=num_1x1_a, kernel_size=1, stride=1, **dd)  # 存在 *args/**kwargs，未转换，需手动确认参数映射;
         self.c3x3_b = BnActConv2d(
-            in_chs=num_1x1_a, out_chs=num_3x3_b, kernel_size=3, stride=self.key_stride, groups=groups, **dd)
+            in_chs=num_1x1_a, out_chs=num_3x3_b, kernel_size=3, stride=self.key_stride, groups=groups, **dd)  # 存在 *args/**kwargs，未转换，需手动确认参数映射;
         if b:
-            self.c1x1_c = CatBnAct(in_chs=num_3x3_b, **dd)
-            self.c1x1_c1 = create_conv2d(num_3x3_b, num_1x1_c, kernel_size=1, **dd)
-            self.c1x1_c2 = create_conv2d(num_3x3_b, inc, kernel_size=1, **dd)
+            self.c1x1_c = CatBnAct(in_chs=num_3x3_b, **dd)  # 存在 *args/**kwargs，未转换，需手动确认参数映射;
+            self.c1x1_c1 = create_conv2d(num_3x3_b, num_1x1_c, kernel_size=1, **dd)  # 存在 *args/**kwargs，未转换，需手动确认参数映射;
+            self.c1x1_c2 = create_conv2d(num_3x3_b, inc, kernel_size=1, **dd)  # 存在 *args/**kwargs，未转换，需手动确认参数映射;
         else:
-            self.c1x1_c = BnActConv2d(in_chs=num_3x3_b, out_chs=num_1x1_c + inc, kernel_size=1, stride=1, **dd)
+            self.c1x1_c = BnActConv2d(in_chs=num_3x3_b, out_chs=num_1x1_c + inc, kernel_size=1, stride=1, **dd)  # 存在 *args/**kwargs，未转换，需手动确认参数映射;
             self.c1x1_c1 = None
             self.c1x1_c2 = None
 
+    # 装饰器 'torch.jit._overload_method' 未在映射表(api_mapping_out_excel.json)中找到，需手动确认;
     @torch.jit._overload_method  # noqa: F811
-    def forward(self, x):
+    def construct(self, x):
         # type: (Tuple[torch.Tensor, torch.Tensor]) -> Tuple[torch.Tensor, torch.Tensor]
         pass
 
+    # 装饰器 'torch.jit._overload_method' 未在映射表(api_mapping_out_excel.json)中找到，需手动确认;
     @torch.jit._overload_method  # noqa: F811
-    def forward(self, x):
+    def construct(self, x):
         # type: (torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]
         pass
 
-    def forward(self, x) -> Tuple[torch.Tensor, torch.Tensor]:
+    def construct(self, x) -> Tuple[ms.Tensor, ms.Tensor]:
         if isinstance(x, tuple):
-            x_in = torch.cat(x, dim=1)
+            x_in = mint.cat(x, dim=1)
         else:
             x_in = x
         if self.c1x1_w_s1 is None and self.c1x1_w_s2 is None:
@@ -164,11 +172,11 @@             out1 = x_in[:, :self.num_1x1_c, :, :]
             out2 = x_in[:, self.num_1x1_c:, :, :]
         resid = x_s1 + out1
-        dense = torch.cat([x_s2, out2], dim=1)
+        dense = mint.cat([x_s2, out2], dim=1)
         return resid, dense
 
 
-class DPN(nn.Module):
+class DPN(msnn.Cell):
     def __init__(
             self,
             k_sec: Tuple[int, ...] = (3, 4, 20, 3),
@@ -209,18 +217,18 @@             stride=2,
             norm_layer=norm_layer,
             **dd,
-        )
-        blocks['conv1_pool'] = nn.MaxPool2d(kernel_size=3, stride=2, padding=1)
+        )  # 存在 *args/**kwargs，未转换，需手动确认参数映射;
+        blocks['conv1_pool'] = nn.MaxPool2d(kernel_size = 3, stride = 2, padding = 1)
         self.feature_info = [dict(num_chs=num_init_features, reduction=2, module='features.conv1_1')]
 
         # conv2
         bw = 64 * bw_factor
         inc = inc_sec[0]
         r = (k_r * bw) // (64 * bw_factor)
-        blocks['conv2_1'] = DualPathBlock(num_init_features, r, r, bw, inc, groups, 'proj', b, **dd)
+        blocks['conv2_1'] = DualPathBlock(num_init_features, r, r, bw, inc, groups, 'proj', b, **dd)  # 存在 *args/**kwargs，未转换，需手动确认参数映射;
         in_chs = bw + 3 * inc
         for i in range(2, k_sec[0] + 1):
-            blocks['conv2_' + str(i)] = DualPathBlock(in_chs, r, r, bw, inc, groups, 'normal', b, **dd)
+            blocks['conv2_' + str(i)] = DualPathBlock(in_chs, r, r, bw, inc, groups, 'normal', b, **dd)  # 存在 *args/**kwargs，未转换，需手动确认参数映射;
             in_chs += inc
         self.feature_info += [dict(num_chs=in_chs, reduction=4, module=f'features.conv2_{k_sec[0]}')]
 
@@ -228,10 +236,10 @@         bw = 128 * bw_factor
         inc = inc_sec[1]
         r = (k_r * bw) // (64 * bw_factor)
-        blocks['conv3_1'] = DualPathBlock(in_chs, r, r, bw, inc, groups, 'down', b, **dd)
+        blocks['conv3_1'] = DualPathBlock(in_chs, r, r, bw, inc, groups, 'down', b, **dd)  # 存在 *args/**kwargs，未转换，需手动确认参数映射;
         in_chs = bw + 3 * inc
         for i in range(2, k_sec[1] + 1):
-            blocks['conv3_' + str(i)] = DualPathBlock(in_chs, r, r, bw, inc, groups, 'normal', b, **dd)
+            blocks['conv3_' + str(i)] = DualPathBlock(in_chs, r, r, bw, inc, groups, 'normal', b, **dd)  # 存在 *args/**kwargs，未转换，需手动确认参数映射;
             in_chs += inc
         self.feature_info += [dict(num_chs=in_chs, reduction=8, module=f'features.conv3_{k_sec[1]}')]
 
@@ -239,10 +247,10 @@         bw = 256 * bw_factor
         inc = inc_sec[2]
         r = (k_r * bw) // (64 * bw_factor)
-        blocks['conv4_1'] = DualPathBlock(in_chs, r, r, bw, inc, groups, 'down', b, **dd)
+        blocks['conv4_1'] = DualPathBlock(in_chs, r, r, bw, inc, groups, 'down', b, **dd)  # 存在 *args/**kwargs，未转换，需手动确认参数映射;
         in_chs = bw + 3 * inc
         for i in range(2, k_sec[2] + 1):
-            blocks['conv4_' + str(i)] = DualPathBlock(in_chs, r, r, bw, inc, groups, 'normal', b, **dd)
+            blocks['conv4_' + str(i)] = DualPathBlock(in_chs, r, r, bw, inc, groups, 'normal', b, **dd)  # 存在 *args/**kwargs，未转换，需手动确认参数映射;
             in_chs += inc
         self.feature_info += [dict(num_chs=in_chs, reduction=16, module=f'features.conv4_{k_sec[2]}')]
 
@@ -250,17 +258,19 @@         bw = 512 * bw_factor
         inc = inc_sec[3]
         r = (k_r * bw) // (64 * bw_factor)
-        blocks['conv5_1'] = DualPathBlock(in_chs, r, r, bw, inc, groups, 'down', b, **dd)
+        blocks['conv5_1'] = DualPathBlock(in_chs, r, r, bw, inc, groups, 'down', b, **dd)  # 存在 *args/**kwargs，未转换，需手动确认参数映射;
         in_chs = bw + 3 * inc
         for i in range(2, k_sec[3] + 1):
-            blocks['conv5_' + str(i)] = DualPathBlock(in_chs, r, r, bw, inc, groups, 'normal', b, **dd)
+            blocks['conv5_' + str(i)] = DualPathBlock(in_chs, r, r, bw, inc, groups, 'normal', b, **dd)  # 存在 *args/**kwargs，未转换，需手动确认参数映射;
             in_chs += inc
         self.feature_info += [dict(num_chs=in_chs, reduction=32, module=f'features.conv5_{k_sec[3]}')]
 
-        blocks['conv5_bn_ac'] = CatBnAct(in_chs, norm_layer=fc_norm_layer, **dd)
+        blocks['conv5_bn_ac'] = CatBnAct(in_chs, norm_layer=fc_norm_layer, **dd)  # 存在 *args/**kwargs，未转换，需手动确认参数映射;
 
         self.num_features = self.head_hidden_size = in_chs
-        self.features = nn.Sequential(blocks)
+        self.features = msnn.SequentialCell([
+            blocks
+        ])
 
         # Using 1x1 conv for the FC layer to allow the extra pooling scheme
         self.global_pool, self.classifier = create_classifier(
@@ -269,10 +279,10 @@             pool_type=global_pool,
             use_conv=True,
             **dd,
-        )
-        self.flatten = nn.Flatten(1) if global_pool else nn.Identity()
-
-    @torch.jit.ignore
+        )  # 存在 *args/**kwargs，未转换，需手动确认参数映射;
+        self.flatten = mint.flatten(1) if global_pool else msnn.Identity()
+
+    @ms.jit
     def group_matcher(self, coarse=False):
         matcher = dict(
             stem=r'^features\.conv1',
@@ -283,19 +293,19 @@         )
         return matcher
 
-    @torch.jit.ignore
+    @ms.jit
     def set_grad_checkpointing(self, enable=True):
         assert not enable, 'gradient checkpointing not supported'
 
-    @torch.jit.ignore
-    def get_classifier(self) -> nn.Module:
+    @ms.jit
+    def get_classifier(self) -> msnn.Cell:
         return self.classifier
 
     def reset_classifier(self, num_classes: int, global_pool: str = 'avg'):
         self.num_classes = num_classes
         self.global_pool, self.classifier = create_classifier(
             self.num_features, self.num_classes, pool_type=global_pool, use_conv=True)
-        self.flatten = nn.Flatten(1) if global_pool else nn.Identity()
+        self.flatten = mint.flatten(1) if global_pool else msnn.Identity()
 
     def forward_features(self, x):
         return self.features(x)
@@ -303,13 +313,13 @@     def forward_head(self, x, pre_logits: bool = False):
         x = self.global_pool(x)
         if self.drop_rate > 0.:
-            x = F.dropout(x, p=self.drop_rate, training=self.training)
+            x = nn.functional.dropout(x, p = self.drop_rate, training = self.training)
         if pre_logits:
             return self.flatten(x)
         x = self.classifier(x)
         return self.flatten(x)
 
-    def forward(self, x):
+    def construct(self, x):
         x = self.forward_features(x)
         x = self.forward_head(x)
         return x
@@ -322,7 +332,7 @@         pretrained,
         feature_cfg=dict(feature_concat=True, flatten_sequential=True),
         **kwargs,
-    )
+    )  # 存在 *args/**kwargs，未转换，需手动确认参数映射;
 
 
 def _cfg(url='', **kwargs):
@@ -355,7 +365,7 @@     model_args = dict(
         small=True, num_init_features=10, k_r=128, groups=32,
         b=True, k_sec=(3, 4, 6, 3), inc_sec=(16, 32, 32, 64), act_layer='silu')
-    return _create_dpn('dpn48b', pretrained=pretrained, **dict(model_args, **kwargs))
+    return _create_dpn('dpn48b', pretrained=pretrained, **dict(model_args, **kwargs))  # 存在 *args/**kwargs，未转换，需手动确认参数映射;
 
 
 @register_model
@@ -363,7 +373,7 @@     model_args = dict(
         small=True, num_init_features=10, k_r=128, groups=32,
         k_sec=(3, 4, 12, 3), inc_sec=(16, 32, 32, 64))
-    return _create_dpn('dpn68', pretrained=pretrained, **dict(model_args, **kwargs))
+    return _create_dpn('dpn68', pretrained=pretrained, **dict(model_args, **kwargs))  # 存在 *args/**kwargs，未转换，需手动确认参数映射;
 
 
 @register_model
@@ -371,7 +381,7 @@     model_args = dict(
         small=True, num_init_features=10, k_r=128, groups=32,
         b=True, k_sec=(3, 4, 12, 3), inc_sec=(16, 32, 32, 64))
-    return _create_dpn('dpn68b', pretrained=pretrained, **dict(model_args, **kwargs))
+    return _create_dpn('dpn68b', pretrained=pretrained, **dict(model_args, **kwargs))  # 存在 *args/**kwargs，未转换，需手动确认参数映射;
 
 
 @register_model
@@ -379,7 +389,7 @@     model_args = dict(
         num_init_features=64, k_r=96, groups=32,
         k_sec=(3, 4, 20, 3), inc_sec=(16, 32, 24, 128))
-    return _create_dpn('dpn92', pretrained=pretrained, **dict(model_args, **kwargs))
+    return _create_dpn('dpn92', pretrained=pretrained, **dict(model_args, **kwargs))  # 存在 *args/**kwargs，未转换，需手动确认参数映射;
 
 
 @register_model
@@ -387,7 +397,7 @@     model_args = dict(
         num_init_features=96, k_r=160, groups=40,
         k_sec=(3, 6, 20, 3), inc_sec=(16, 32, 32, 128))
-    return _create_dpn('dpn98', pretrained=pretrained, **dict(model_args, **kwargs))
+    return _create_dpn('dpn98', pretrained=pretrained, **dict(model_args, **kwargs))  # 存在 *args/**kwargs，未转换，需手动确认参数映射;
 
 
 @register_model
@@ -395,7 +405,7 @@     model_args = dict(
         num_init_features=128, k_r=160, groups=40,
         k_sec=(4, 8, 28, 3), inc_sec=(16, 32, 32, 128))
-    return _create_dpn('dpn131', pretrained=pretrained, **dict(model_args, **kwargs))
+    return _create_dpn('dpn131', pretrained=pretrained, **dict(model_args, **kwargs))  # 存在 *args/**kwargs，未转换，需手动确认参数映射;
 
 
 @register_model
@@ -403,4 +413,4 @@     model_args = dict(
         num_init_features=128, k_r=200, groups=50,
         k_sec=(4, 8, 20, 3), inc_sec=(20, 64, 64, 128))
-    return _create_dpn('dpn107', pretrained=pretrained, **dict(model_args, **kwargs))
+    return _create_dpn('dpn107', pretrained=pretrained, **dict(model_args, **kwargs))  # 存在 *args/**kwargs，未转换，需手动确认参数映射;
