--- pytorch+++ mindspore@@ -1,12 +1,15 @@+import mindspore as ms
+import mindspore.nn as msnn
+import mindspore.ops as msops
+import mindspore.mint as mint
+from mindspore.mint import nn, ops
 """ Res2Net and Res2NeXt
 Adapted from Official Pytorch impl at: https://github.com/gasvn/Res2Net/
 Paper: `Res2Net: A New Multi-scale Backbone Architecture` - https://arxiv.org/abs/1904.01169
 """
 import math
 from typing import Optional, Type
-
-import torch
-import torch.nn as nn
+# import torch.nn as nn
 
 from timm.data import IMAGENET_DEFAULT_MEAN, IMAGENET_DEFAULT_STD
 from ._builder import build_model_with_cfg
@@ -16,7 +19,7 @@ __all__ = []
 
 
-class Bottle2neck(nn.Module):
+class Bottle2neck(msnn.Cell):
     """ Res2Net/Res2NeXT Bottleneck
     Adapted from https://github.com/gasvn/Res2Net/blob/master/res2net.py
     """
@@ -68,11 +71,11 @@                 **dd,
             ))
             bns.append(norm_layer(width, **dd))
-        self.convs = nn.ModuleList(convs)
-        self.bns = nn.ModuleList(bns)
+        self.convs = msnn.CellList(convs)
+        self.bns = msnn.CellList(bns)
         if self.is_first:
             # FIXME this should probably have count_include_pad=False, but hurts original weights
-            self.pool = nn.AvgPool2d(kernel_size=3, stride=stride, padding=1)
+            self.pool = nn.AvgPool2d(kernel_size = 3, stride = stride, padding = 1)
         else:
             self.pool = None
 
@@ -85,16 +88,16 @@ 
     def zero_init_last(self):
         if getattr(self.bn3, 'weight', None) is not None:
-            nn.init.zeros_(self.bn3.weight)
-
-    def forward(self, x):
+            nn.init.zeros_(self.bn3.weight)  # 'torch.nn.init.zeros_' 未在映射表(api_mapping_out_excel.json)中找到，需手动确认;
+
+    def construct(self, x):
         shortcut = x
 
         out = self.conv1(x)
         out = self.bn1(out)
         out = self.relu(out)
 
-        spx = torch.split(out, self.width, 1)
+        spx = mint.split(out, self.width, 1)
         spo = []
         sp = spx[0]  # redundant, for torchscript
         for i, (conv, bn) in enumerate(zip(self.convs, self.bns)):
@@ -111,7 +114,7 @@                 spo.append(self.pool(spx[-1]))
             else:
                 spo.append(spx[-1])
-        out = torch.cat(spo, 1)
+        out = mint.cat(spo, 1)
 
         out = self.conv3(out)
         out = self.bn3(out)
