--- pytorch+++ mindspore@@ -1,3 +1,8 @@+import mindspore as ms
+import mindspore.nn as msnn
+import mindspore.ops as msops
+import mindspore.mint as mint
+from mindspore.mint import nn, ops
 """SHViT
 SHViT: Single-Head Vision Transformer with Memory Efficient Macro Design
 Code: https://github.com/ysj9909/SHViT
@@ -13,9 +18,8 @@ """
 from typing import Any, Dict, List, Optional, Set, Tuple, Type, Union
 
-import torch
-import torch.nn as nn
-import torch.nn.functional as F
+# import torch
+# import torch.nn as nn
 
 from timm.data import IMAGENET_DEFAULT_MEAN, IMAGENET_DEFAULT_STD
 from timm.layers import GroupNorm1, SqueezeExcite, SelectAdaptivePool2d, LayerType, trunc_normal_
@@ -27,28 +31,28 @@ __all__ = ['SHViT']
 
 
-class Residual(nn.Module):
-    def __init__(self, m: nn.Module):
+class Residual(msnn.Cell):
+    def __init__(self, m: msnn.Cell):
         super().__init__()
         self.m = m
 
-    def forward(self, x: torch.Tensor) -> torch.Tensor:
+    def construct(self, x: ms.Tensor) -> ms.Tensor:
         return x + self.m(x)
 
     @torch.no_grad()
-    def fuse(self) -> nn.Module:
+    def fuse(self) -> msnn.Cell:
         if isinstance(self.m, Conv2dNorm):
             m = self.m.fuse()
             assert(m.groups == m.in_channels)
-            identity = torch.ones(m.weight.shape[0], m.weight.shape[1], 1, 1)
-            identity = F.pad(identity, [1,1,1,1])
+            identity = mint.ones(size = (m.weight.shape[0], m.weight.shape[1], 1, 1))
+            identity = nn.functional.pad(identity, [1,1,1,1])
             m.weight += identity.to(m.weight.device)
             return m
         else:
             return self
 
 
-class Conv2dNorm(nn.Sequential):
+class Conv2dNorm(msnn.SequentialCell):
     def __init__(
             self,
             in_channels: int,
@@ -66,8 +70,8 @@         self.add_module('c', nn.Conv2d(
             in_channels, out_channels, kernel_size, stride, padding, bias=False, **dd, **kwargs))
         self.add_module('bn', nn.BatchNorm2d(out_channels, **dd))
-        nn.init.constant_(self.bn.weight, bn_weight_init)
-        nn.init.constant_(self.bn.bias, 0)
+        nn.init.constant_(self.bn.weight, bn_weight_init)  # 'torch.nn.init.constant_' 未在映射表(api_mapping_out_excel.json)中找到，需手动确认;
+        nn.init.constant_(self.bn.bias, 0)  # 'torch.nn.init.constant_' 未在映射表(api_mapping_out_excel.json)中找到，需手动确认;
 
     @torch.no_grad()
     def fuse(self) -> nn.Conv2d:
@@ -76,22 +80,13 @@         w = c.weight * w[:, None, None, None]
         b = bn.bias - bn.running_mean * bn.weight / (bn.running_var + bn.eps) ** 0.5
         m = nn.Conv2d(
-            in_channels=w.size(1) * self.c.groups,
-            out_channels=w.size(0),
-            kernel_size=w.shape[2:],
-            stride=self.c.stride,
-            padding=self.c.padding,
-            dilation=self.c.dilation,
-            groups=self.c.groups,
-            device=c.weight.device,
-            dtype=c.weight.dtype,
-        )
+            in_channels = w.size(1) * self.c.groups, out_channels = w.size(0), kernel_size = w.shape[2:], stride = self.c.stride, padding = self.c.padding, dilation = self.c.dilation, groups = self.c.groups, dtype = c.weight.dtype)  # 'torch.nn.Conv2d':没有对应的mindspore参数 'device' (position 9);
         m.weight.data.copy_(w)
         m.bias.data.copy_(b)
         return m
 
 
-class NormLinear(nn.Sequential):
+class NormLinear(msnn.SequentialCell):
     def __init__(
             self,
             in_features: int,
@@ -107,7 +102,7 @@         self.add_module('l', nn.Linear(in_features, out_features, bias=bias, **dd))
         trunc_normal_(self.l.weight, std=std)
         if bias:
-            nn.init.constant_(self.l.bias, 0)
+            nn.init.constant_(self.l.bias, 0)  # 'torch.nn.init.constant_' 未在映射表(api_mapping_out_excel.json)中找到，需手动确认;
 
     @torch.no_grad()
     def fuse(self) -> nn.Linear:
@@ -119,18 +114,18 @@             b = b @ self.l.weight.T
         else:
             b = (l.weight @ b[:, None]).view(-1) + self.l.bias
-        m = nn.Linear(w.size(1), w.size(0), device=l.weight.device, dtype=l.weight.dtype)
+        m = nn.Linear(w.size(1), w.size(0), dtype = l.weight.dtype)  # 'torch.nn.Linear':没有对应的mindspore参数 'device' (position 3);
         m.weight.data.copy_(w)
         m.bias.data.copy_(b)
         return m
 
 
-class PatchMerging(nn.Module):
+class PatchMerging(msnn.Cell):
     def __init__(
             self,
             dim: int,
             out_dim: int,
-            act_layer: Type[nn.Module] = nn.ReLU,
+            act_layer: Type[msnn.Cell] = nn.ReLU,
             device=None,
             dtype=None,
     ):
@@ -144,7 +139,7 @@         self.se = SqueezeExcite(hid_dim, 0.25, **dd)
         self.conv3 = Conv2dNorm(hid_dim, out_dim, **dd)
 
-    def forward(self, x: torch.Tensor) -> torch.Tensor:
+    def construct(self, x: ms.Tensor) -> ms.Tensor:
         x = self.conv1(x)
         x = self.act1(x)
         x = self.conv2(x)
@@ -154,12 +149,12 @@         return x
 
 
-class FFN(nn.Module):
+class FFN(msnn.Cell):
     def __init__(
             self,
             dim: int,
             embed_dim: int,
-            act_layer: Type[nn.Module] = nn.ReLU,
+            act_layer: Type[msnn.Cell] = nn.ReLU,
             device=None,
             dtype=None,
     ):
@@ -169,22 +164,22 @@         self.act = act_layer()
         self.pw2 = Conv2dNorm(embed_dim, dim, bn_weight_init=0, **dd)
 
-    def forward(self, x: torch.Tensor) -> torch.Tensor:
+    def construct(self, x: ms.Tensor) -> ms.Tensor:
         x = self.pw1(x)
         x = self.act(x)
         x = self.pw2(x)
         return x
 
 
-class SHSA(nn.Module):
+class SHSA(msnn.Cell):
     """Single-Head Self-Attention"""
     def __init__(
             self,
             dim: int,
             qk_dim: int,
             pdim: int,
-            norm_layer: Type[nn.Module] = GroupNorm1,
-            act_layer: Type[nn.Module] = nn.ReLU,
+            norm_layer: Type[msnn.Cell] = GroupNorm1,
+            act_layer: Type[msnn.Cell] = nn.ReLU,
             device=None,
             dtype=None,
     ):
@@ -198,32 +193,35 @@         self.pre_norm = norm_layer(pdim, **dd)
 
         self.qkv = Conv2dNorm(pdim, qk_dim * 2 + pdim, **dd)
-        self.proj = nn.Sequential(act_layer(), Conv2dNorm(dim, dim, bn_weight_init=0, **dd))
-
-    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        self.proj = msnn.SequentialCell([
+            act_layer(),
+            Conv2dNorm(dim, dim, bn_weight_init=0, **dd)
+        ])
+
+    def construct(self, x: ms.Tensor) -> ms.Tensor:
         B, _, H, W = x.shape
-        x1, x2 = torch.split(x, [self.pdim, self.dim - self.pdim], dim = 1)
+        x1, x2 = mint.split(x, [self.pdim, self.dim - self.pdim], dim = 1)
         x1 = self.pre_norm(x1)
         qkv = self.qkv(x1)
-        q, k, v = torch.split(qkv, [self.qk_dim, self.qk_dim, self.pdim], dim=1)
+        q, k, v = mint.split(qkv, [self.qk_dim, self.qk_dim, self.pdim], dim = 1)
         q, k, v = q.flatten(2), k.flatten(2), v.flatten(2)
 
         attn = (q.transpose(-2, -1) @ k) * self.scale
         attn = attn.softmax(dim=-1)
         x1 = (v @ attn.transpose(-2, -1)).reshape(B, self.pdim, H, W)
-        x = self.proj(torch.cat([x1, x2], dim = 1))
-        return x
-
-
-class BasicBlock(nn.Module):
+        x = self.proj(mint.cat([x1, x2], dim = 1))
+        return x
+
+
+class BasicBlock(msnn.Cell):
     def __init__(
             self,
             dim: int,
             qk_dim: int,
             pdim: int,
             type: str,
-            norm_layer: Type[nn.Module] = GroupNorm1,
-            act_layer: Type[nn.Module] = nn.ReLU,
+            norm_layer: Type[msnn.Cell] = GroupNorm1,
+            act_layer: Type[msnn.Cell] = nn.ReLU,
             device=None,
             dtype=None,
     ):
@@ -233,17 +231,17 @@         if type == "s":
             self.mixer = Residual(SHSA(dim, qk_dim, pdim, norm_layer, act_layer, **dd))
         else:
-            self.mixer = nn.Identity()
+            self.mixer = msnn.Identity()
         self.ffn = Residual(FFN(dim, int(dim * 2), **dd))
 
-    def forward(self, x: torch.Tensor) -> torch.Tensor:
+    def construct(self, x: ms.Tensor) -> ms.Tensor:
         x = self.conv(x)
         x = self.mixer(x)
         x = self.ffn(x)
         return x
 
 
-class StageBlock(nn.Module):
+class StageBlock(msnn.Cell):
     def __init__(
             self,
             prev_dim: int,
@@ -252,27 +250,30 @@             pdim: int,
             type: str,
             depth: int,
-            norm_layer: Type[nn.Module] = GroupNorm1,
-            act_layer: Type[nn.Module] = nn.ReLU,
+            norm_layer: Type[msnn.Cell] = GroupNorm1,
+            act_layer: Type[msnn.Cell] = nn.ReLU,
             device=None,
             dtype=None,
     ):
         dd = {'device': device, 'dtype': dtype}
         super().__init__()
         self.grad_checkpointing = False
-        self.downsample = nn.Sequential(
+        self.downsample = msnn.SequentialCell(
+            [
             Residual(Conv2dNorm(prev_dim, prev_dim, 3, 1, 1, groups=prev_dim, **dd)),
             Residual(FFN(prev_dim, int(prev_dim * 2), act_layer, **dd)),
             PatchMerging(prev_dim, dim, act_layer, **dd),
             Residual(Conv2dNorm(dim, dim, 3, 1, 1, groups=dim, **dd)),
-            Residual(FFN(dim, int(dim * 2), act_layer, **dd)),
-        ) if prev_dim != dim else nn.Identity()
-
-        self.blocks = nn.Sequential(*[
+            Residual(FFN(dim, int(dim * 2), act_layer, **dd))
+        ]) if prev_dim != dim else msnn.Identity()
+
+        self.blocks = msnn.SequentialCell([
+            [
             BasicBlock(dim, qk_dim, pdim, type, norm_layer, act_layer, **dd) for _ in range(depth)
+        ]
         ])
 
-    def forward(self, x: torch.Tensor) -> torch.Tensor:
+    def construct(self, x: ms.Tensor) -> ms.Tensor:
         x = self.downsample(x)
         if self.grad_checkpointing and not torch.jit.is_scripting():
             x = checkpoint_seq(self.blocks, x)
@@ -281,7 +282,7 @@         return x
 
 
-class SHViT(nn.Module):
+class SHViT(msnn.Cell):
     def __init__(
             self,
             in_chans: int = 3,
@@ -293,8 +294,8 @@             depth: Tuple[int, int, int] = (1, 2, 3),
             types: Tuple[str, str, str] = ("s", "s", "s"),
             drop_rate: float = 0.,
-            norm_layer: Type[nn.Module] = GroupNorm1,
-            act_layer: Type[nn.Module] = nn.ReLU,
+            norm_layer: Type[msnn.Cell] = GroupNorm1,
+            act_layer: Type[msnn.Cell] = nn.ReLU,
             device=None,
             dtype=None,
     ):
@@ -306,7 +307,8 @@ 
         # Patch embedding
         stem_chs = embed_dim[0]
-        self.patch_embed = nn.Sequential(
+        self.patch_embed = msnn.SequentialCell(
+            [
             Conv2dNorm(in_chans, stem_chs // 8, 3, 2, 1, **dd),
             act_layer(),
             Conv2dNorm(stem_chs // 8, stem_chs // 4, 3, 2, 1, **dd),
@@ -314,7 +316,7 @@             Conv2dNorm(stem_chs // 4, stem_chs // 2, 3, 2, 1, **dd),
             act_layer(),
             Conv2dNorm(stem_chs // 2, stem_chs, 3, 2, 1, **dd)
-        )
+        ])
 
         # Build SHViT blocks
         stages = []
@@ -333,13 +335,15 @@             ))
             prev_chs = embed_dim[i]
             self.feature_info.append(dict(num_chs=prev_chs, reduction=2**(i+4), module=f'stages.{i}'))
-        self.stages = nn.Sequential(*stages)
+        self.stages = msnn.SequentialCell([
+            stages
+        ])
 
         # Classifier head
         self.num_features = self.head_hidden_size = embed_dim[-1]
         self.global_pool = SelectAdaptivePool2d(pool_type=global_pool)
-        self.flatten = nn.Flatten(1) if global_pool else nn.Identity()  # don't flatten if pooling disabled
-        self.head = NormLinear(self.head_hidden_size, num_classes, **dd) if num_classes > 0 else nn.Identity()
+        self.flatten = mint.flatten(1) if global_pool else msnn.Identity()  # don't flatten if pooling disabled
+        self.head = NormLinear(self.head_hidden_size, num_classes, **dd) if num_classes > 0 else msnn.Identity()
 
     @torch.jit.ignore
     def no_weight_decay(self) -> Set:
@@ -362,25 +366,25 @@             s.grad_checkpointing = enable
 
     @torch.jit.ignore
-    def get_classifier(self) -> nn.Module:
+    def get_classifier(self) -> msnn.Cell:
         return self.head.l
 
     def reset_classifier(self, num_classes: int, global_pool: str = 'avg'):
         self.num_classes = num_classes
         # cannot meaningfully change pooling of efficient head after creation
         self.global_pool = SelectAdaptivePool2d(pool_type=global_pool)
-        self.flatten = nn.Flatten(1) if global_pool else nn.Identity()  # don't flatten if pooling disabled
-        self.head = NormLinear(self.head_hidden_size, num_classes) if num_classes > 0 else nn.Identity()
+        self.flatten = mint.flatten(1) if global_pool else msnn.Identity()  # don't flatten if pooling disabled
+        self.head = NormLinear(self.head_hidden_size, num_classes) if num_classes > 0 else msnn.Identity()
 
     def forward_intermediates(
             self,
-            x: torch.Tensor,
+            x: ms.Tensor,
             indices: Optional[Union[int, List[int]]] = None,
             norm: bool = False,
             stop_early: bool = False,
             output_fmt: str = 'NCHW',
             intermediates_only: bool = False,
-    ) -> Union[List[torch.Tensor], Tuple[torch.Tensor, List[torch.Tensor]]]:
+    ) -> Union[List[ms.Tensor], Tuple[ms.Tensor, List[ms.Tensor]]]:
         """ Forward features that returns intermediates.
 
         Args:
@@ -428,19 +432,19 @@             self.reset_classifier(0, '')
         return take_indices
 
-    def forward_features(self, x: torch.Tensor) -> torch.Tensor:
+    def forward_features(self, x: ms.Tensor) -> ms.Tensor:
         x = self.patch_embed(x)
         x = self.stages(x)
         return x
 
-    def forward_head(self, x: torch.Tensor, pre_logits: bool = False) -> torch.Tensor:
+    def forward_head(self, x: ms.Tensor, pre_logits: bool = False) -> ms.Tensor:
         x = self.global_pool(x)
         x = self.flatten(x)
         if self.drop_rate > 0.:
-            x = F.dropout(x, p=self.drop_rate, training=self.training)
+            x = nn.functional.dropout(x, p = self.drop_rate, training = self.training)
         return x if pre_logits else self.head(x)
 
-    def forward(self, x: torch.Tensor) -> torch.Tensor:
+    def construct(self, x: ms.Tensor) -> ms.Tensor:
         x = self.forward_features(x)
         x = self.forward_head(x)
         return x
@@ -459,7 +463,7 @@         fuse_children(self)
 
 
-def checkpoint_filter_fn(state_dict: Dict[str, torch.Tensor], model: nn.Module) -> Dict[str, torch.Tensor]:
+def checkpoint_filter_fn(state_dict: Dict[str, ms.Tensor], model: msnn.Cell) -> Dict[str, ms.Tensor]:
     state_dict = state_dict.get('model', state_dict)
 
     # out_dict = {}
